From 5920fa9bb2fbaf4400dcba99bc326139a4c5b85d Mon Sep 17 00:00:00 2001
From: Tommy Chen <tommyyl.chen@mediatek.com>
Date: Fri, 29 Aug 2025 15:26:46 +0800
Subject: [PATCH 1/3] GENIO: drm/mediatek: Add DVO and eDP driver for MT8391

Add DVO and eDP drivers for customers verifying the eDP MT8391
Weston is able to show on the eDP panel

Signed-off-by: Tommy Chen <tommyyl.chen@mediatek.com>

squash
---
 drivers/gpu/drm/mediatek/Makefile           |   1 +
 drivers/gpu/drm/mediatek/mtk_disp_drv.h     |   3 +
 drivers/gpu/drm/mediatek/mtk_dp.c           | 460 ++++++++--
 drivers/gpu/drm/mediatek/mtk_dp_reg.h       | 110 ++-
 drivers/gpu/drm/mediatek/mtk_drm_ddp_comp.c |  10 +-
 drivers/gpu/drm/mediatek/mtk_drm_drv.c      |  23 +-
 drivers/gpu/drm/mediatek/mtk_drm_drv.h      |   1 +
 drivers/gpu/drm/mediatek/mtk_dvo.c          | 885 ++++++++++++++++++++
 drivers/gpu/drm/mediatek/mtk_dvo_regs.h     | 158 ++++
 9 files changed, 1563 insertions(+), 88 deletions(-)
 create mode 100644 drivers/gpu/drm/mediatek/mtk_dvo.c
 create mode 100644 drivers/gpu/drm/mediatek/mtk_dvo_regs.h

diff --git a/drivers/gpu/drm/mediatek/Makefile b/drivers/gpu/drm/mediatek/Makefile
index 880abbdee049..c03348008ed4 100644
--- a/drivers/gpu/drm/mediatek/Makefile
+++ b/drivers/gpu/drm/mediatek/Makefile
@@ -17,6 +17,7 @@ mediatek-drm-y := mtk_disp_aal.o \
 		  mtk_dsi.o \
 		  mtk_dpi.o \
 		  mtk_ethdr.o \
+		  mtk_dvo.o \
 		  mtk_mdp_rdma.o
 
 obj-$(CONFIG_DRM_MEDIATEK) += mediatek-drm.o
diff --git a/drivers/gpu/drm/mediatek/mtk_disp_drv.h b/drivers/gpu/drm/mediatek/mtk_disp_drv.h
index 12382f8290b9..173c690813af 100644
--- a/drivers/gpu/drm/mediatek/mtk_disp_drv.h
+++ b/drivers/gpu/drm/mediatek/mtk_disp_drv.h
@@ -30,6 +30,9 @@ void mtk_ccorr_config(struct device *dev, unsigned int w,
 void mtk_ccorr_start(struct device *dev);
 void mtk_ccorr_stop(struct device *dev);
 
+void mtk_dvo_start(struct device *dev);
+void mtk_dvo_stop(struct device *dev);
+
 void mtk_color_bypass_shadow(struct device *dev);
 int mtk_color_clk_enable(struct device *dev);
 void mtk_color_clk_disable(struct device *dev);
diff --git a/drivers/gpu/drm/mediatek/mtk_dp.c b/drivers/gpu/drm/mediatek/mtk_dp.c
index 07fa3ad8f73c..2cf475f5d29f 100644
--- a/drivers/gpu/drm/mediatek/mtk_dp.c
+++ b/drivers/gpu/drm/mediatek/mtk_dp.c
@@ -35,6 +35,9 @@
 
 #include "mtk_dp_reg.h"
 
+#define EDP_VIDEO_UNMUTE		0x22
+#define EDP_VIDEO_UNMUTE_VAL		0xfefd
+#define MTK_SIP_DP_CONTROL		(0x82000523 | 0x40000000)
 #define MTK_DP_SIP_CONTROL_AARCH32	MTK_SIP_SMC_CMD(0x523)
 #define MTK_DP_SIP_ATF_EDP_VIDEO_UNMUTE	(BIT(0) | BIT(5))
 #define MTK_DP_SIP_ATF_VIDEO_UNMUTE	BIT(5)
@@ -66,6 +69,11 @@ enum {
 	MTK_DP_CAL_MAX,
 };
 
+enum {
+	MTK_EDP_SOC_TYPE_NONE  = 0,
+	MTK_EDP_SOC_TYPE_MT8189,
+};
+
 struct mtk_dp_train_info {
 	bool sink_ssc;
 	bool cable_plugged_in;
@@ -124,6 +132,7 @@ struct mtk_dp {
 	struct platform_device *phy_dev;
 	struct phy *phy;
 	struct regmap *regs;
+	struct regmap *phy_regs;
 	struct timer_list debounce_timer;
 
 	/* For audio */
@@ -144,6 +153,7 @@ struct mtk_dp_data {
 	bool audio_supported;
 	bool audio_pkt_in_hblank_area;
 	u16 audio_m_div2_bit;
+	u32 edp_ver;
 };
 
 static const struct mtk_dp_efuse_fmt mt8188_dp_efuse_fmt[MTK_DP_CAL_MAX] = {
@@ -403,6 +413,14 @@ static struct regmap_config mtk_dp_regmap_config = {
 	.name = "mtk-dp-registers",
 };
 
+static const struct regmap_config mtk_edp_phy_regmap_config = {
+	.reg_bits = 32,
+	.val_bits = 32,
+	.reg_stride = 4,
+	.max_register = SEC_OFFSET + 0x90,
+	.name = "mtk-edp-phy-registers",
+};
+
 static struct mtk_dp *mtk_dp_from_bridge(struct drm_bridge *b)
 {
 	return container_of(b, struct mtk_dp, bridge);
@@ -515,9 +533,14 @@ static void mtk_dp_set_msa(struct mtk_dp *mtk_dp)
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_315C,
 			   vm->hsync_len,
 			   PGEN_HSYNC_PULSE_WIDTH_DP_ENC0_P0_MASK);
-	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3160,
-			   vm->hback_porch + vm->hsync_len,
-			   PGEN_HFDE_START_DP_ENC0_P0_MASK);
+	if (mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3160,
+				   vm->hback_porch + vm->hsync_len + vm->hfront_porch,
+				   PGEN_HFDE_START_DP_ENC0_P0_MASK);
+	else
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3160,
+				   vm->hback_porch + vm->hsync_len,
+				   PGEN_HFDE_START_DP_ENC0_P0_MASK);
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3164,
 			   vm->hactive,
 			   PGEN_HFDE_ACTIVE_WIDTH_DP_ENC0_P0_MASK);
@@ -532,9 +555,14 @@ static void mtk_dp_set_msa(struct mtk_dp *mtk_dp)
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3170,
 			   vm->vsync_len,
 			   PGEN_VSYNC_PULSE_WIDTH_DP_ENC0_P0_MASK);
-	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3174,
-			   vm->vback_porch + vm->vsync_len,
-			   PGEN_VFDE_START_DP_ENC0_P0_MASK);
+	if (mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3174,
+				   vm->vback_porch + vm->vsync_len +  vm->vfront_porch,
+				   PGEN_VFDE_START_DP_ENC0_P0_MASK);
+	else
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3174,
+				   vm->vback_porch + vm->vsync_len,
+				   PGEN_VFDE_START_DP_ENC0_P0_MASK);
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3178,
 			   vm->vactive,
 			   PGEN_VFDE_ACTIVE_WIDTH_DP_ENC0_P0_MASK);
@@ -613,7 +641,8 @@ static void mtk_dp_setup_encoder(struct mtk_dp *mtk_dp)
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_3364,
 			   FIFO_READ_START_POINT_DP_ENC1_P0_VAL << 12,
 			   FIFO_READ_START_POINT_DP_ENC1_P0_MASK);
-	mtk_dp_write(mtk_dp, MTK_DP_ENC1_P0_3368, DP_ENC1_P0_3368_VAL);
+	if (!mtk_dp->data->edp_ver)
+		mtk_dp_write(mtk_dp, MTK_DP_ENC1_P0_3368, DP_ENC1_P0_3368_VAL);
 }
 
 static void mtk_dp_pg_enable(struct mtk_dp *mtk_dp, bool enable)
@@ -982,6 +1011,37 @@ static void mtk_dp_set_swing_pre_emphasis(struct mtk_dp *mtk_dp, int lane_num,
 	mtk_dp_update_bits(mtk_dp, MTK_DP_TOP_SWING_EMP,
 			   preemphasis << (DP_TX0_PRE_EMPH_SHIFT + lane_shift),
 			   DP_TX0_PRE_EMPH_MASK << lane_shift);
+
+	if (mtk_dp->data->edp_ver && mtk_dp->phy_regs) {
+		/* set swing and pre */
+		switch (lane_num) {
+		case 0:
+		case 1:
+		case 2:
+		case 3:
+			if (mtk_dp->data->edp_ver == MTK_EDP_SOC_TYPE_MT8189) {
+				regmap_update_bits(mtk_dp->phy_regs,
+					   PHYD_DIG_DRV_FORCE_LANE_8189(lane_num),
+					   EDP_TX_LN_VOLT_SWING_VAL_8189_FLDMASK |
+					   EDP_TX_LN_PRE_EMPH_VAL_8189_FLDMASK,
+					   swing_val << 1 |
+					   preemphasis << 5);
+				regmap_update_bits(mtk_dp->phy_regs,
+					   PHYD_DIG_DRV_FORCE_LANE_8189(lane_num),
+					   EDP_TX_LN_VOLT_SWING_EN_8189 |
+					   EDP_TX_LN_PRE_EMPH_EN_8189,
+					   EDP_TX_LN_VOLT_SWING_EN_8189 |
+					   EDP_TX_LN_PRE_EMPH_EN_8189);
+			}
+			break;
+		default:
+			break;
+		}
+		if (preemphasis != 0)
+			mtk_dp_update_bits(mtk_dp, RG_DSI_DEM_EN,
+					   DSI_DE_EMPHASIS_ENABLE,
+					   DSI_DE_EMPHASIS_ENABLE);
+	}
 }
 
 static void mtk_dp_reset_swing_pre_emphasis(struct mtk_dp *mtk_dp)
@@ -1015,14 +1075,27 @@ static u32 mtk_dp_swirq_get_clear(struct mtk_dp *mtk_dp)
 
 static u32 mtk_dp_hwirq_get_clear(struct mtk_dp *mtk_dp)
 {
-	u32 irq_status = (mtk_dp_read(mtk_dp, MTK_DP_TRANS_P0_3418) &
-			  IRQ_STATUS_DP_TRANS_P0_MASK) >> 12;
+	u32 irq_status;
 
-	if (irq_status) {
-		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
-				   irq_status, IRQ_CLR_DP_TRANS_P0_MASK);
-		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
-				   0, IRQ_CLR_DP_TRANS_P0_MASK);
+	if (mtk_dp->data->edp_ver) {
+		irq_status = (mtk_dp_read(mtk_dp, MTK_DP_AUX_P0_3608));
+		if (irq_status) {
+			mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3668,
+					   irq_status, irq_status);
+			mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3668,
+					   0, irq_status);
+		}
+	} else {
+		irq_status = (mtk_dp_read(mtk_dp, MTK_DP_TRANS_P0_3418) &
+			      IRQ_STATUS_DP_TRANS_P0_MASK) >> 12;
+
+		if (irq_status) {
+			mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
+					   irq_status,
+					   IRQ_CLR_DP_TRANS_P0_MASK);
+			mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
+					   0, IRQ_CLR_DP_TRANS_P0_MASK);
+		}
 	}
 
 	return irq_status;
@@ -1030,16 +1103,64 @@ static u32 mtk_dp_hwirq_get_clear(struct mtk_dp *mtk_dp)
 
 static void mtk_dp_hwirq_enable(struct mtk_dp *mtk_dp, bool enable)
 {
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
-			   enable ? 0 :
-			   IRQ_MASK_DP_TRANS_P0_DISC_IRQ |
-			   IRQ_MASK_DP_TRANS_P0_CONN_IRQ |
-			   IRQ_MASK_DP_TRANS_P0_INT_IRQ,
-			   IRQ_MASK_DP_TRANS_P0_MASK);
+	if (mtk_dp->data->edp_ver) {
+		if (enable)
+			mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3660, 0x0,
+					   HPD_CONNECT_EVENT |
+					   HPD_INTERRUPT_EVENT |
+					   HPD_DISCONNECT_EVENT);
+		else
+			mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3660,
+					   AUX_DP_TX_INT_3660_VAL,
+					   AUX_DP_TX_INT_3660_MASK);
+	} else {
+			mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3418,
+					   enable ? 0 :
+					   IRQ_MASK_DP_TRANS_P0_CONN_IRQ |
+					   IRQ_MASK_DP_TRANS_P0_INT_IRQ,
+					   IRQ_MASK_DP_TRANS_P0_MASK);
+	}
 }
 
 static void mtk_dp_initialize_settings(struct mtk_dp *mtk_dp)
 {
+	if (mtk_dp->data->edp_ver) {
+		mtk_dp_update_bits(mtk_dp, REG_3F04_DP_ENC_P0_3, 0,
+				   FRAME_START_MARKER_0_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F08_DP_ENC_P0_3,
+				   FRAME_START_MARKER_1_DP_ENC_P0_3,
+				   FRAME_START_MARKER_1_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F0C_DP_ENC_P0_3,
+				   FRAME_END_MARKER_0_DP_ENC_P0_3,
+				   FRAME_END_MARKER_0_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F10_DP_ENC_P0_3,
+				   FRAME_END_MARKER_1_DP_ENC_P0_3,
+				   FRAME_END_MARKER_1_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_33C0, 0,
+				   SDP_TESTBUS_SEL_DP_ENC_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_33C0,
+				   SDP_TESTBUS_SEL_BIT4_DP_ENC,
+				   SDP_TESTBUS_SEL_BIT4_DP_ENC_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_33C4,
+				   DP_TX_ENCODER_TESTBUS_SEL_DP_ENC,
+				   DP_TX_ENCODER_TESTBUS_SEL_DP_ENC_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F28_DP_ENC_P0_3,
+				   DP_TX_SDP_PSR_AS_TESTBUS,
+				   DP_TX_SDP_PSR_AS_TESTBUS_MASK);
+		mtk_dp_update_bits(mtk_dp, DP_TX_TOP_RESET_AND_PROBE,
+				   RG_SW_RST,
+				   RG_SW_RST_MASK);
+		mtk_dp_update_bits(mtk_dp, DP_TX_TOP_RESET_AND_PROBE,
+				   RG_PROBE_LOW_SEL,
+				   RG_PROBE_LOW_SEL_MASK);
+		mtk_dp_update_bits(mtk_dp, DP_TX_TOP_RESET_AND_PROBE,
+				   RG_PROBE_LOW_HIGH_SWAP,
+				   RG_PROBE_LOW_HIGH_SWAP_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TOP_IRQ_MASK,
+				   ENCODER_IRQ_MSK | TRANS_IRQ_MSK,
+				   ENCODER_IRQ_MSK | TRANS_IRQ_MSK);
+	}
+
 	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_342C,
 			   XTAL_FREQ_DP_TRANS_P0_DEFAULT,
 			   XTAL_FREQ_DP_TRANS_P0_MASK);
@@ -1058,27 +1179,40 @@ static void mtk_dp_initialize_settings(struct mtk_dp *mtk_dp)
 static void mtk_dp_initialize_hpd_detect_settings(struct mtk_dp *mtk_dp)
 {
 	u32 val;
-	/* Debounce threshold */
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
-			   8, HPD_DEB_THD_DP_TRANS_P0_MASK);
 
-	val = (HPD_INT_THD_DP_TRANS_P0_LOWER_500US |
-	       HPD_INT_THD_DP_TRANS_P0_UPPER_1100US) << 4;
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
-			   val, HPD_INT_THD_DP_TRANS_P0_MASK);
+	if (mtk_dp->data->edp_ver) {
+		mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_364C,
+				   HPD_INT_THD_FLDMASK_VAL << 4,
+				   HPD_INT_THD_FLDMASK);
+	} else {
+		/* Debounce threshold */
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
+				   8, HPD_DEB_THD_DP_TRANS_P0_MASK);
 
-	/*
-	 * Connect threshold 1.5ms + 5 x 0.1ms = 2ms
-	 * Disconnect threshold 1.5ms + 5 x 0.1ms = 2ms
-	 */
-	val = (5 << 8) | (5 << 12);
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
-			   val,
-			   HPD_DISC_THD_DP_TRANS_P0_MASK |
-			   HPD_CONN_THD_DP_TRANS_P0_MASK);
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3430,
-			   HPD_INT_THD_ECO_DP_TRANS_P0_HIGH_BOUND_EXT,
-			   HPD_INT_THD_ECO_DP_TRANS_P0_MASK);
+		val = (HPD_INT_THD_DP_TRANS_P0_LOWER_500US |
+		       HPD_INT_THD_DP_TRANS_P0_UPPER_1100US) << 4;
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
+				   val, HPD_INT_THD_DP_TRANS_P0_MASK);
+
+		/*
+		 * Connect threshold 1.5ms + 5 x 0.1ms = 2ms
+		 * Disconnect threshold 1.5ms + 5 x 0.1ms = 2ms
+		 */
+		val = (5 << 8) | (5 << 12);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3410,
+				   val,
+				   HPD_DISC_THD_DP_TRANS_P0_MASK |
+				   HPD_CONN_THD_DP_TRANS_P0_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_3430,
+				   HPD_INT_THD_ECO_DP_TRANS_P0_HIGH_BOUND_EXT,
+				   HPD_INT_THD_ECO_DP_TRANS_P0_MASK);
+	}
+}
+
+static void mtk_edp_phyd_wait_aux_ldo_ready(struct mtk_dp *mtk_dp,
+					    unsigned long wait_us)
+{
+	mdelay(wait_us/1000);
 }
 
 static void mtk_dp_initialize_aux_settings(struct mtk_dp *mtk_dp)
@@ -1089,6 +1223,11 @@ static void mtk_dp_initialize_aux_settings(struct mtk_dp *mtk_dp)
 			   AUX_TIMEOUT_THR_AUX_TX_P0_MASK);
 	mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3658,
 			   0, AUX_TX_OV_EN_AUX_TX_P0_MASK);
+
+	if (mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_36A0,
+				   DP_TX_INIT_MASK_15_TO_2,
+				   DP_TX_INIT_MASK_15_TO_2_MASK);
 	/* 25 for 26M */
 	mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3634,
 			   AUX_TX_OVER_SAMPLE_RATE_FOR_26M << 8,
@@ -1105,6 +1244,23 @@ static void mtk_dp_initialize_aux_settings(struct mtk_dp *mtk_dp)
 	mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_3690,
 			   RX_REPLY_COMPLETE_MODE_AUX_TX_P0,
 			   RX_REPLY_COMPLETE_MODE_AUX_TX_P0);
+
+	if (mtk_dp->data->edp_ver) {
+		/*Con Thd = 1.5ms+Vx0.1ms*/
+		mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_367C,
+				   HPD_CONN_THD_AUX_TX_P0_FLDMASK_POS << 6,
+				   HPD_CONN_THD_AUX_TX_P0_FLDMASK);
+		/*DisCon Thd = 1.5ms+Vx0.1ms*/
+		mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_37A0,
+				   HPD_DISC_THD_AUX_TX_P0_FLDMASK_POS << 4,
+				   HPD_DISC_THD_AUX_TX_P0_FLDMASK);
+		mtk_dp_update_bits(mtk_dp, REG_3FF8_DP_ENC_P0_3,
+				   XTAL_FREQ_FOR_PSR_DP_ENC_P0_3_VALUE << 9,
+				   XTAL_FREQ_FOR_PSR_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_AUX_P0_366C,
+				   XTAL_FREQ_DP_TX_AUX_366C_VALUE << 8,
+				   XTAL_FREQ_DP_TX_AUX_366C_MASK);
+	}
 }
 
 static void mtk_dp_initialize_digital_settings(struct mtk_dp *mtk_dp)
@@ -1116,15 +1272,61 @@ static void mtk_dp_initialize_digital_settings(struct mtk_dp *mtk_dp)
 			   BS2BS_MODE_DP_ENC1_P0_VAL << 12,
 			   BS2BS_MODE_DP_ENC1_P0_MASK);
 
+	if (mtk_dp->data->edp_ver) {
+		/* dp I-mode enable */
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3000,
+				   DP_I_MODE_ENABLE, DP_I_MODE_ENABLE);
+		/*symbol_cnt_reset */
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3000,
+				   REG_BS_SYMBOL_CNT_RESET,
+				   REG_BS_SYMBOL_CNT_RESET);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_3368,
+				   VIDEO_SRAM_FIFO_CNT_RESET_SEL_DP_ENC1_P0,
+				   VIDEO_SRAM_FIFO_CNT_RESET_SEL_MASK);
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC1_P0_3368,
+				   BS_FOLLOW_SEL_DP_ENC0_P0,
+				   BS_FOLLOW_SEL_DP_ENC0_P0);
+		/*[5:0]video sram start address*/
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_303C,
+				   SRAM_START_READ_THRD_DP_ENC0_P0_VALUE,
+				   SRAM_START_READ_THRD_DP_ENC0_P0_MASK);
+		/* reg_psr_patgen_avt_en disable psr pattern */
+		mtk_dp_update_bits(mtk_dp, REG_3F80_DP_ENC_P0_3,
+				   0, PSR_PATGEN_AVT_EN_FLDMASK);
+		/* phy D enable */
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   ALL_POWER_ON,
+				   PHY_PWR_STATE_OW_VALUE_DP_ENC_P0_3_MASK);
+		mtk_edp_phyd_wait_aux_ldo_ready(mtk_dp, 100000);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   0, PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3FF8_DP_ENC_P0_3,
+				   PHY_STATE_W_1_DP_ENC_P0_3,
+				   PHY_STATE_W_1_DP_ENC_P0_3_MASK);
+		/* reg_dvo_on_ow_en */
+		mtk_dp_update_bits(mtk_dp, REG_3FF8_DP_ENC_P0_3,
+				   DVO_ON_W_1_FLDMASK,
+				   DVO_ON_W_1_FLDMASK);
+	}
 	/* dp tx encoder reset all sw */
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3004,
 			   DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0,
 			   DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0);
-
+	if (mtk_dp->data->edp_ver) {
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3004, 0,
+				   DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0);
+		mtk_dp_update_bits(mtk_dp, REG_3FF8_DP_ENC_P0_3,
+				   PHY_STATE_RESET_ALL_VALUE,
+				   PHY_STATE_RESET_ALL_MASK);
+	}
 	/* Wait for sw reset to complete */
 	usleep_range(1000, 5000);
-	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3004,
-			   0, DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0);
+	if (!mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3004,
+				   0, DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0);
 }
 
 static void mtk_dp_digital_sw_reset(struct mtk_dp *mtk_dp)
@@ -1141,6 +1343,18 @@ static void mtk_dp_digital_sw_reset(struct mtk_dp *mtk_dp)
 
 static void mtk_dp_set_lanes(struct mtk_dp *mtk_dp, int lanes)
 {
+	if (mtk_dp->data->edp_ver) {
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   BIAS_POWER_ON,
+				   PHY_PWR_STATE_OW_VALUE_DP_ENC_P0_3_MASK);
+		mtk_edp_phyd_wait_aux_ldo_ready(mtk_dp, 100000);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   0, PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+	}
+
 	mtk_dp_update_bits(mtk_dp, MTK_DP_TRANS_P0_35F0,
 			   lanes == 0 ? 0 : DP_TRANS_DUMMY_RW_0,
 			   DP_TRANS_DUMMY_RW_0_MASK);
@@ -1158,8 +1372,7 @@ static void mtk_dp_get_calibration_data(struct mtk_dp *mtk_dp)
 	u32 *cal_data = mtk_dp->cal_data;
 	u32 *buf;
 	int i;
-	size_t len = 0;
-
+	size_t len;
 	cell = nvmem_cell_get(dev, "dp_calibration_data");
 	if (IS_ERR(cell)) {
 		dev_warn(dev, "Failed to get nvmem cell dp_calibration_data\n");
@@ -1169,17 +1382,21 @@ static void mtk_dp_get_calibration_data(struct mtk_dp *mtk_dp)
 	buf = (u32 *)nvmem_cell_read(cell, &len);
 	nvmem_cell_put(cell);
 
-	if (IS_ERR(buf) || ((len / sizeof(u32)) != 4)) {
+	if (IS_ERR(buf)) {
 		dev_warn(dev, "Failed to read nvmem_cell_read\n");
-
-		if (!IS_ERR(buf))
-			kfree(buf);
-
 		goto use_default_val;
 	}
-
+	/* The cell length is in bytes. Convert it to be compatible with u32 buffer. */
+	len /= sizeof(u32);
 	for (i = 0; i < MTK_DP_CAL_MAX; i++) {
 		fmt = &mtk_dp->data->efuse_fmt[i];
+		if (fmt->idx >= len) {
+			dev_warn(mtk_dp->dev,
+				"Out-of-bound efuse data access, fmt idx = %d, buf len = %zu\n",
+				fmt->idx, len);
+			kfree(buf);
+			goto use_default_val;
+		}
 		cal_data[i] = (buf[fmt->idx] >> fmt->shift) & fmt->mask;
 
 		if (cal_data[i] < fmt->min_val || cal_data[i] > fmt->max_val) {
@@ -1247,15 +1464,19 @@ static int mtk_dp_phy_configure(struct mtk_dp *mtk_dp,
 			.ssc = mtk_dp->train_info.sink_ssc,
 		}
 	};
-
-	mtk_dp_update_bits(mtk_dp, MTK_DP_TOP_PWR_STATE, DP_PWR_STATE_BANDGAP,
-			   DP_PWR_STATE_MASK);
-
+	if (!mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, MTK_DP_TOP_PWR_STATE,
+				   DP_PWR_STATE_BANDGAP, DP_PWR_STATE_MASK);
 	ret = phy_configure(mtk_dp->phy, &phy_opts);
 	if (ret)
 		return ret;
 
 	mtk_dp_set_calibration_data(mtk_dp);
+	/* Turn on phy power after phy configure */
+	if (mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, REG_3FF8_DP_ENC_P0_3,
+				   PHY_STATE_W_1_DP_ENC_P0_3,
+				   PHY_STATE_W_1_DP_ENC_P0_3_MASK);
 	mtk_dp_update_bits(mtk_dp, MTK_DP_TOP_PWR_STATE,
 			   DP_PWR_STATE_BANDGAP_TPLL_LANE, DP_PWR_STATE_MASK);
 
@@ -1307,15 +1528,21 @@ static void mtk_dp_video_mute(struct mtk_dp *mtk_dp, bool enable)
 	struct arm_smccc_res res;
 	u32 val = VIDEO_MUTE_SEL_DP_ENC0_P0 |
 		  (enable ? VIDEO_MUTE_SW_DP_ENC0_P0 : 0);
+	u32 smmc_para;
 
 	mtk_dp_update_bits(mtk_dp, MTK_DP_ENC0_P0_3000,
 			   val,
 			   VIDEO_MUTE_SEL_DP_ENC0_P0 |
 			   VIDEO_MUTE_SW_DP_ENC0_P0);
-
-	arm_smccc_smc(MTK_DP_SIP_CONTROL_AARCH32,
-		      mtk_dp->data->smc_cmd, enable,
-		      0, 0, 0, 0, 0, &res);
+	if (mtk_dp->data->edp_ver) {
+		smmc_para = (EDP_VIDEO_UNMUTE << 16) | enable;
+		arm_smccc_smc(MTK_SIP_DP_CONTROL, EDP_VIDEO_UNMUTE, enable,
+			      smmc_para, EDP_VIDEO_UNMUTE_VAL, 0, 0, 0, &res);
+	} else {
+		arm_smccc_smc(MTK_DP_SIP_CONTROL_AARCH32,
+			      mtk_dp->data->smc_cmd, enable,
+			      0, 0, 0, 0, 0, &res);
+	}
 
 	dev_dbg(mtk_dp->dev, "smc cmd: 0x%x, p1: %s, ret: 0x%lx-0x%lx\n",
 		mtk_dp->data->smc_cmd, enable ? "enable" : "disable", res.a0, res.a1);
@@ -1391,7 +1618,16 @@ static void mtk_dp_power_enable(struct mtk_dp *mtk_dp)
 static void mtk_dp_power_disable(struct mtk_dp *mtk_dp)
 {
 	mtk_dp_write(mtk_dp, MTK_DP_TOP_PWR_STATE, 0);
-
+	if (mtk_dp->data->edp_ver) {
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3,
+				   PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   ALL_POWER_OFF,
+				   PHY_PWR_STATE_OW_VALUE_DP_ENC_P0_3_MASK);
+		mtk_dp_update_bits(mtk_dp, REG_3F44_DP_ENC_P0_3,
+				   0, PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK);
+	}
 	mtk_dp_update_bits(mtk_dp, MTK_DP_0034,
 			   DA_CKM_CKTX0_EN_FORCE_EN, DA_CKM_CKTX0_EN_FORCE_EN);
 
@@ -1405,7 +1641,10 @@ static void mtk_dp_initialize_priv_data(struct mtk_dp *mtk_dp)
 {
 	bool plugged_in = (mtk_dp->bridge.type == DRM_MODE_CONNECTOR_eDP);
 
-	mtk_dp->train_info.link_rate = DP_LINK_BW_5_4;
+	if (mtk_dp->data->edp_ver)
+		mtk_dp->train_info.link_rate = DP_LINK_BW_8_1;
+	else
+		mtk_dp->train_info.link_rate = DP_LINK_BW_5_4;
 	mtk_dp->train_info.lane_count = mtk_dp->max_lanes;
 	mtk_dp->train_info.cable_plugged_in = plugged_in;
 
@@ -1950,6 +2189,10 @@ static void mtk_dp_init_port(struct mtk_dp *mtk_dp)
 	mtk_dp_initialize_hpd_detect_settings(mtk_dp);
 
 	mtk_dp_digital_sw_reset(mtk_dp);
+	if (mtk_dp->data->edp_ver)
+		mtk_dp_update_bits(mtk_dp, EDP_TX_TOP_CLKGEN_0,
+				   EDP_TX_TOP_CLKGEN_REST_VALUE,
+				   EDP_TX_TOP_CLKGEN_REST_MASK);
 }
 
 static irqreturn_t mtk_dp_hpd_event_thread(int hpd, void *dev)
@@ -2000,25 +2243,42 @@ static irqreturn_t mtk_dp_hpd_event(int hpd, void *dev)
 
 	spin_lock_irqsave(&mtk_dp->irq_thread_lock, flags);
 
-	if (irq_status & MTK_DP_HPD_INTERRUPT)
-		mtk_dp->irq_thread_handle |= MTK_DP_THREAD_HPD_EVENT;
+	if (mtk_dp->data->edp_ver) {
+		if (irq_status & MTK_EDP_HPD_INTERRUPT)
+			mtk_dp->irq_thread_handle |= MTK_DP_THREAD_HPD_EVENT;
+		/* Cable state is changed. */
+		if (irq_status != MTK_EDP_HPD_INTERRUPT) {
+			mtk_dp->irq_thread_handle |= MTK_DP_THREAD_CABLE_STATE_CHG;
+			cable_sta_chg = true;
+		}
+		spin_unlock_irqrestore(&mtk_dp->irq_thread_lock, flags);
+		if (cable_sta_chg) {
+			if (!!(mtk_dp_read(mtk_dp, MTK_DP_AUX_P0_364C) &
+				HPD_STATUS_DP_AUX_TX_P0_MASK))
+				mtk_dp->train_info.cable_plugged_in = true;
+			else
+				mtk_dp->train_info.cable_plugged_in = false;
+		}
+	} else {
+		if (irq_status & MTK_DP_HPD_INTERRUPT)
+			mtk_dp->irq_thread_handle |= MTK_DP_THREAD_HPD_EVENT;
 
-	/* Cable state is changed. */
-	if (irq_status != MTK_DP_HPD_INTERRUPT) {
-		mtk_dp->irq_thread_handle |= MTK_DP_THREAD_CABLE_STATE_CHG;
-		cable_sta_chg = true;
-	}
+		/* Cable state is changed. */
+		if (irq_status != MTK_DP_HPD_INTERRUPT) {
+			mtk_dp->irq_thread_handle |= MTK_DP_THREAD_CABLE_STATE_CHG;
+			cable_sta_chg = true;
+		}
 
-	spin_unlock_irqrestore(&mtk_dp->irq_thread_lock, flags);
+		spin_unlock_irqrestore(&mtk_dp->irq_thread_lock, flags);
 
-	if (cable_sta_chg) {
-		if (!!(mtk_dp_read(mtk_dp, MTK_DP_TRANS_P0_3414) &
-		       HPD_DB_DP_TRANS_P0_MASK))
-			mtk_dp->train_info.cable_plugged_in = true;
-		else
-			mtk_dp->train_info.cable_plugged_in = false;
+		if (cable_sta_chg) {
+			if (!!(mtk_dp_read(mtk_dp, MTK_DP_TRANS_P0_3414) &
+			       HPD_DB_DP_TRANS_P0_MASK))
+				mtk_dp->train_info.cable_plugged_in = true;
+			else
+				mtk_dp->train_info.cable_plugged_in = false;
+		}
 	}
-
 	return IRQ_WAKE_THREAD;
 }
 
@@ -2028,7 +2288,12 @@ static int mtk_dp_wait_hpd_asserted(struct drm_dp_aux *mtk_aux, unsigned long wa
 	u32 val;
 	int ret;
 
-	ret = regmap_read_poll_timeout(mtk_dp->regs, MTK_DP_TRANS_P0_3414,
+	if (mtk_dp->data->edp_ver)
+		ret = regmap_read_poll_timeout(mtk_dp->regs, MTK_DP_AUX_P0_364C,
+					       val, !!(val & HPD_STATUS_DP_AUX_TX_P0_MASK),
+					       wait_us / 100, wait_us);
+	else
+		ret = regmap_read_poll_timeout(mtk_dp->regs, MTK_DP_TRANS_P0_3414,
 				       val, !!(val & HPD_DB_DP_TRANS_P0_MASK),
 				       wait_us / 100, wait_us);
 	if (ret) {
@@ -2054,6 +2319,7 @@ static int mtk_dp_dt_parse(struct mtk_dp *mtk_dp,
 	struct device *dev = &pdev->dev;
 	int ret;
 	void __iomem *base;
+	void __iomem *phy_base;
 	u32 linkrate;
 	int len;
 
@@ -2065,6 +2331,14 @@ static int mtk_dp_dt_parse(struct mtk_dp *mtk_dp,
 	if (IS_ERR(mtk_dp->regs))
 		return PTR_ERR(mtk_dp->regs);
 
+	phy_base = devm_platform_ioremap_resource(pdev, 1);
+	if (!IS_ERR_OR_NULL(phy_base)) {
+		mtk_dp->phy_regs = devm_regmap_init_mmio(dev, phy_base,
+					&mtk_edp_phy_regmap_config);
+		if (IS_ERR(mtk_dp->phy_regs))
+			mtk_dp->phy_regs = NULL;
+	}
+
 	endpoint = of_graph_get_endpoint_by_regs(pdev->dev.of_node, 1, -1);
 	len = of_property_count_elems_of_size(endpoint,
 					      "data-lanes", sizeof(u32));
@@ -2166,7 +2440,7 @@ static const struct drm_edid *mtk_dp_edid_read(struct drm_bridge *bridge,
 		if (ret < 0)
 			ret = 0;
 		if (ret)
-			kfree(sads);
+		kfree(sads);
 		audio_caps->sad_count = ret;
 
 		/*
@@ -2674,17 +2948,30 @@ static int mtk_dp_register_phy(struct mtk_dp *mtk_dp)
 {
 	struct device *dev = mtk_dp->dev;
 
-	mtk_dp->phy_dev = platform_device_register_data(dev, "mediatek-dp-phy",
+	if (mtk_dp->data->edp_ver == MTK_EDP_SOC_TYPE_MT8189) {
+		if (mtk_dp->phy_regs)
+			mtk_dp->phy_dev = platform_device_register_data(dev,
+						"mediatek-edp-phy",
+						PLATFORM_DEVID_AUTO,
+						&mtk_dp->phy_regs,
+						sizeof(struct regmap *));
+	} else {
+		mtk_dp->phy_dev = platform_device_register_data(dev, "mediatek-dp-phy",
 							PLATFORM_DEVID_AUTO,
 							&mtk_dp->regs,
 							sizeof(struct regmap *));
+	}
+
 	if (IS_ERR(mtk_dp->phy_dev))
 		return dev_err_probe(dev, PTR_ERR(mtk_dp->phy_dev),
 				     "Failed to create device mediatek-dp-phy\n");
 
 	mtk_dp_get_calibration_data(mtk_dp);
 
-	mtk_dp->phy = devm_phy_get(&mtk_dp->phy_dev->dev, "dp");
+	if (mtk_dp->data->edp_ver)
+		mtk_dp->phy = devm_phy_get(&mtk_dp->phy_dev->dev, "edp");
+	else
+		mtk_dp->phy = devm_phy_get(&mtk_dp->phy_dev->dev, "dp");
 	if (IS_ERR(mtk_dp->phy)) {
 		platform_device_unregister(mtk_dp->phy_dev);
 		return dev_err_probe(dev, PTR_ERR(mtk_dp->phy), "Failed to get phy\n");
@@ -2912,6 +3199,15 @@ static const struct mtk_dp_data mt8195_dp_data = {
 	.audio_m_div2_bit = MT8195_AUDIO_M_CODE_MULT_DIV_SEL_DP_ENC0_P0_DIV_2,
 };
 
+static const struct mtk_dp_data mt8189_edp_data = {
+	.bridge_type = DRM_MODE_CONNECTOR_eDP,
+	.smc_cmd = MTK_DP_SIP_ATF_EDP_VIDEO_UNMUTE,
+	.efuse_fmt = mt8195_edp_efuse_fmt,
+	.audio_supported = false,
+	.audio_m_div2_bit = MT8195_AUDIO_M_CODE_MULT_DIV_SEL_DP_ENC0_P0_DIV_2,
+	.edp_ver = MTK_EDP_SOC_TYPE_MT8189,
+};
+
 static const struct of_device_id mtk_dp_of_match[] = {
 	{
 		.compatible = "mediatek,mt8188-edp-tx",
@@ -2929,6 +3225,10 @@ static const struct of_device_id mtk_dp_of_match[] = {
 		.compatible = "mediatek,mt8195-dp-tx",
 		.data = &mt8195_dp_data,
 	},
+	{
+		.compatible = "mediatek,mt8189-edp-tx",
+		.data = &mt8189_edp_data,
+	},
 	{},
 };
 MODULE_DEVICE_TABLE(of, mtk_dp_of_match);
@@ -2950,4 +3250,4 @@ MODULE_AUTHOR("Markus Schneider-Pargmann <msp@baylibre.com>");
 MODULE_AUTHOR("Bo-Chen Chen <rex-bc.chen@mediatek.com>");
 MODULE_DESCRIPTION("MediaTek DisplayPort Driver");
 MODULE_LICENSE("GPL");
-MODULE_SOFTDEP("pre: phy_mtk_dp");
+MODULE_SOFTDEP("pre: phy_mtk_dp");
\ No newline at end of file
diff --git a/drivers/gpu/drm/mediatek/mtk_dp_reg.h b/drivers/gpu/drm/mediatek/mtk_dp_reg.h
index 8bf3f8686834..d03738eb85ae 100644
--- a/drivers/gpu/drm/mediatek/mtk_dp_reg.h
+++ b/drivers/gpu/drm/mediatek/mtk_dp_reg.h
@@ -11,12 +11,20 @@
 #define MTK_DP_HPD_DISCONNECT		BIT(1)
 #define MTK_DP_HPD_CONNECT		BIT(2)
 #define MTK_DP_HPD_INTERRUPT		BIT(3)
+#define MTK_EDP_HPD_INTERRUPT		BIT(2)
 
 /* offset: 0x0 */
 #define DP_PHY_GLB_BIAS_GEN_00		0x0
 #define RG_XTP_GLB_BIAS_INTR_CTRL		GENMASK(20, 16)
 #define DP_PHY_GLB_DPAUX_TX		0x8
 #define RG_CKM_PT0_CKTX_IMPSEL			GENMASK(23, 20)
+#define DP_TX_TOP_RESET_AND_PROBE	0x2020
+#define RG_SW_RST_MASK				GENMASK(7, 0)
+#define RG_SW_RST				0xff
+#define RG_PROBE_LOW_SEL_MASK			GENMASK(18, 16)
+#define RG_PROBE_LOW_SEL			BIT(16)
+#define RG_PROBE_LOW_HIGH_SWAP_MASK		BIT(23)
+#define RG_PROBE_LOW_HIGH_SWAP			BIT(23)
 #define MTK_DP_0034			0x34
 #define DA_XTP_GLB_CKDET_EN_FORCE_VAL		BIT(15)
 #define DA_XTP_GLB_CKDET_EN_FORCE_EN		BIT(14)
@@ -51,6 +59,13 @@
 #define RG_XTP_GLB_CKDET_EN			BIT(1)
 #define RG_DPAUX_RX_EN				BIT(0)
 
+/* offset: PHY_OFFSET */
+#define DP_PHY_DIG_GLB_STATUS_0		0x146c
+#define RGS_BG_CORE_EN_READY			BIT(0)
+#define RGS_AUX_LDO_EN_READY			BIT(1)
+#define MTK_EDP_PWR_DOMAIN		0x0074
+#define EDP_PWR_RST				BIT(0)
+#define EDP_PWR_CLK_DIS				BIT(4)
 /* offset: TOP_OFFSET (0x2000) */
 #define MTK_DP_TOP_PWR_STATE		0x2000
 #define DP_PWR_STATE_MASK			GENMASK(1, 0)
@@ -71,18 +86,26 @@
 #define DP_TX3_PRE_EMPH_MASK			GENMASK(27, 26)
 #define MTK_DP_TOP_RESET_AND_PROBE	0x2020
 #define SW_RST_B_PHYD				BIT(4)
+#define DP_TX_TOP_SOFT_PROBE		0x2024
 #define MTK_DP_TOP_IRQ_MASK		0x202c
+#define ENCODER_IRQ_MSK				BIT(0)
+#define TRANS_IRQ_MSK				BIT(1)
 #define IRQ_MASK_AUX_TOP_IRQ			BIT(2)
+#define IRQ_OUT_HIGH_ACTIVE_FLDMASK		0x100
 #define MTK_DP_TOP_MEM_PD		0x2038
 #define MEM_ISO_EN				BIT(0)
 #define FUSE_SEL				BIT(2)
-
+#define EDP_TX_TOP_CLKGEN_0		0x2074
+#define EDP_TX_TOP_CLKGEN_REST_MASK		0xf
+#define EDP_TX_TOP_CLKGEN_REST_VALUE		0xf
 /* offset: ENC0_OFFSET (0x3000) */
 #define MTK_DP_ENC0_P0_3000			0x3000
 #define LANE_NUM_DP_ENC0_P0_MASK			GENMASK(1, 0)
 #define VIDEO_MUTE_SW_DP_ENC0_P0			BIT(2)
 #define VIDEO_MUTE_SEL_DP_ENC0_P0			BIT(3)
 #define ENHANCED_FRAME_EN_DP_ENC0_P0			BIT(4)
+#define DP_I_MODE_ENABLE				BIT(6)
+#define REG_BS_SYMBOL_CNT_RESET				BIT(7)
 #define MTK_DP_ENC0_P0_3004			0x3004
 #define VIDEO_M_CODE_SEL_DP_ENC0_P0_MASK		BIT(8)
 #define DP_TX_ENCODER_4P_RESET_SW_DP_ENC0_P0		BIT(9)
@@ -122,6 +145,7 @@
 #define VIDEO_SOURCE_SEL_DP_ENC0_P0_MASK		BIT(11)
 #define MTK_DP_ENC0_P0_303C			0x303c
 #define SRAM_START_READ_THRD_DP_ENC0_P0_MASK		GENMASK(5, 0)
+#define SRAM_START_READ_THRD_DP_ENC0_P0_VALUE		0x8
 #define VIDEO_COLOR_DEPTH_DP_ENC0_P0_MASK		GENMASK(10, 8)
 #define VIDEO_COLOR_DEPTH_DP_ENC0_P0_16BIT		(0 << 8)
 #define VIDEO_COLOR_DEPTH_DP_ENC0_P0_12BIT		(1 << 8)
@@ -225,6 +249,7 @@
 #define FIFO_READ_START_POINT_DP_ENC1_P0_MASK		GENMASK(15, 12)
 #define MTK_DP_ENC1_P0_3368			0x3368
 #define VIDEO_SRAM_FIFO_CNT_RESET_SEL_DP_ENC1_P0	BIT(0)
+#define VIDEO_SRAM_FIFO_CNT_RESET_SEL_MASK		GENMASK(1, 0)
 #define VIDEO_STABLE_CNT_THRD_DP_ENC1_P0		BIT(4)
 #define SDP_DP13_EN_DP_ENC1_P0				BIT(8)
 #define BS2BS_MODE_DP_ENC1_P0				BIT(12)
@@ -234,11 +259,18 @@
 							 VIDEO_STABLE_CNT_THRD_DP_ENC1_P0 | \
 							 SDP_DP13_EN_DP_ENC1_P0 | \
 							 BS2BS_MODE_DP_ENC1_P0)
+#define BS_FOLLOW_SEL_DP_ENC0_P0			BIT(15)
 
 #define MTK_DP_ENC1_P0_3374			0x3374
 #define SDP_ASP_INSERT_IN_HBLANK_DP_ENC1_P0_MASK	BIT(12)
 #define SDP_DOWN_ASP_CNT_INIT_DP_ENC1_P0_MASK		GENMASK(11, 0)
-
+#define MTK_DP_ENC1_P0_33C0			0x33c0
+#define SDP_TESTBUS_SEL_DP_ENC_MASK			0xf000
+#define SDP_TESTBUS_SEL_BIT4_DP_ENC_MASK		0x80
+#define SDP_TESTBUS_SEL_BIT4_DP_ENC			0x80
+#define MTK_DP_ENC1_P0_33C4			0x33c4
+#define DP_TX_ENCODER_TESTBUS_SEL_DP_ENC_MASK		GENMASK(6, 5)
+#define DP_TX_ENCODER_TESTBUS_SEL_DP_ENC		BIT(5)
 #define MTK_DP_ENC1_P0_33F4			0x33f4
 #define DP_ENC_DUMMY_RW_1_AUDIO_RST_EN			BIT(0)
 #define DP_ENC_DUMMY_RW_1				BIT(9)
@@ -297,9 +329,14 @@
 #define DP_TRANS_DUMMY_RW_0_MASK				GENMASK(3, 2)
 
 /* offset: AUX_OFFSET (0x3600) */
+#define MTK_DP_AUX_P0_3608			0x3608
+#define HPD_CONNECT_EVENT				BIT(0)
+#define HPD_INTERRUPT_EVENT				BIT(2)
+#define HPD_DISCONNECT_EVENT				BIT(10)
 #define MTK_DP_AUX_P0_360C			0x360c
 #define AUX_TIMEOUT_THR_AUX_TX_P0_MASK			GENMASK(12, 0)
 #define AUX_TIMEOUT_THR_AUX_TX_P0_VAL			0x1595
+#define AUX_TIMEOUT_THR_AUX_TX_P0_VAL_1			0x1d0c
 #define MTK_DP_AUX_P0_3614			0x3614
 #define AUX_RX_UI_CNT_THR_AUX_TX_P0_MASK		GENMASK(6, 0)
 #define AUX_RX_UI_CNT_THR_AUX_FOR_26M			13
@@ -345,19 +382,86 @@
 #define MCU_REQUEST_ADDRESS_LSB_AUX_TX_P0_MASK		GENMASK(15, 0)
 #define MTK_DP_AUX_P0_364C			0x364c
 #define MCU_REQUEST_ADDRESS_MSB_AUX_TX_P0_MASK		GENMASK(3, 0)
+#define HPD_INT_THD_FLDMASK_VAL				0x32
+#define HPD_INT_THD_FLDMASK				GENMASK(9, 4)
+#define HPD_STATUS_DP_AUX_TX_P0_MASK			BIT(15)
 #define MTK_DP_AUX_P0_3650			0x3650
 #define MCU_REQ_DATA_NUM_AUX_TX_P0_MASK			GENMASK(15, 12)
 #define PHY_FIFO_RST_AUX_TX_P0_MASK			BIT(9)
 #define MCU_ACK_TRAN_COMPLETE_AUX_TX_P0			BIT(8)
 #define MTK_DP_AUX_P0_3658			0x3658
 #define AUX_TX_OV_EN_AUX_TX_P0_MASK			BIT(0)
+#define MTK_DP_AUX_P0_3660			0x3660
+#define AUX_DP_TX_INT_3660_VAL			0xffff
+#define AUX_DP_TX_INT_3660_MASK				GENMASK(15, 0)
+#define MTK_DP_AUX_P0_3668			0x3668
+#define MTK_DP_AUX_P0_366C			0x366c
+#define XTAL_FREQ_DP_TX_AUX_366C_VALUE			0x68
+#define XTAL_FREQ_DP_TX_AUX_366C_MASK			0xff00
+#define MTK_DP_AUX_P0_367C			0x367c
+#define HPD_CONN_THD_AUX_TX_P0_FLDMASK_POS		5
+#define HPD_CONN_THD_AUX_TX_P0_FLDMASK			0x3c0
 #define MTK_DP_AUX_P0_3690			0x3690
 #define RX_REPLY_COMPLETE_MODE_AUX_TX_P0		BIT(8)
+#define MTK_DP_AUX_P0_36A0			0x36a0
+#define DP_TX_INIT_MASK_15_TO_2_MASK			0xfffc
+#define DP_TX_INIT_MASK_15_TO_2				0xfffc
 #define MTK_DP_AUX_P0_3704			0x3704
 #define AUX_TX_FIFO_WDATA_NEW_MODE_T_AUX_TX_P0_MASK	BIT(1)
 #define AUX_TX_FIFO_NEW_MODE_EN_AUX_TX_P0		BIT(2)
 #define MTK_DP_AUX_P0_3708			0x3708
+#define MTK_DP_AUX_P0_37A0			0x37a0
+#define HPD_DISC_THD_AUX_TX_P0_FLDMASK_POS		5
+#define HPD_DISC_THD_AUX_TX_P0_FLDMASK			0xf0
 #define MTK_DP_AUX_P0_37C8			0x37c8
 #define MTK_ATOP_EN_AUX_TX_P0				BIT(0)
 
-#endif /*_MTK_DP_REG_H_*/
+/* offset info queue */
+#define REG_3F04_DP_ENC_P0_3			0x3f04
+#define FRAME_START_MARKER_0_DP_ENC_P0_3_MASK		GENMASK(15, 0)
+#define REG_3F08_DP_ENC_P0_3			0x3f08
+#define FRAME_START_MARKER_1_DP_ENC_P0_3_MASK		BIT(3)
+#define FRAME_START_MARKER_1_DP_ENC_P0_3		BIT(3)
+#define REG_3F0C_DP_ENC_P0_3			0x3f0C
+#define FRAME_END_MARKER_0_DP_ENC_P0_3_MASK		BIT(1)
+#define FRAME_END_MARKER_0_DP_ENC_P0_3			BIT(1)
+#define REG_3F10_DP_ENC_P0_3			0x3f10
+#define FRAME_END_MARKER_1_DP_ENC_P0_3_MASK		BIT(3)
+#define FRAME_END_MARKER_1_DP_ENC_P0_3			BIT(3)
+#define REG_3F28_DP_ENC_P0_3			0x3f28
+#define DP_TX_SDP_PSR_AS_TESTBUS_MASK			GENMASK(5, 2)
+#define DP_TX_SDP_PSR_AS_TESTBUS			(0xa << 2)
+#define REG_3F40_DP_ENC_P0_3			0x3f40
+#define REG_3F44_DP_ENC_P0_3			0x3f44
+#define PHY_PWR_STATE_OW_EN_DP_ENC_P0_3			BIT(2)
+#define PHY_PWR_STATE_OW_EN_DP_ENC_P0_3_MASK		BIT(2)
+#define ALL_POWER_OFF					(0x0 << 3)
+#define BIAS_POWER_ON					(0x1 << 3)
+#define BIAS_AND_PLL_POWER_ON				(0x2 << 3)
+#define ALL_POWER_ON					(0x3 << 3)
+#define PHY_PWR_STATE_OW_VALUE_DP_ENC_P0_3_MASK		GENMASK(4, 3)
+#define REG_3F80_DP_ENC_P0_3			0x3f80
+#define PSR_PATGEN_AVT_EN_FLDMASK			BIT(5)
+#define REG_3FF8_DP_ENC_P0_3			0x3ff8
+#define PHY_STATE_RESET_ALL_MASK			GENMASK(7, 0)
+#define PHY_STATE_RESET_ALL_VALUE			0xff
+#define PHY_STATE_W_1_DP_ENC_P0_3			BIT(6)
+#define PHY_STATE_W_1_DP_ENC_P0_3_MASK			BIT(6)
+#define DVO_ON_W_1_FLDMASK				BIT(5)
+#define XTAL_FREQ_FOR_PSR_DP_ENC_P0_3_VALUE		25
+#define XTAL_FREQ_FOR_PSR_DP_ENC_P0_3_MASK		GENMASK(13, 9)
+#define DISP_EDPTX_PWR_RST_B				(1 << 0)
+#define DISP_EDPTX_PWR_CLK_DIS				(1 << 4)
+
+/* offset: PHYD_OFFSET (0x1000) */
+#define PHYD_DIG_DRV_FORCE_LANE(lane)		(0x1030 + 0x100 * (lane))
+#define EDP_TX_LN_VOLT_SWING_VAL_FLDMASK	0x6
+#define EDP_TX_LN_PRE_EMPH_VAL_FLDMASK		0x18
+#define PHYD_DIG_DRV_FORCE_LANE_8189(lane)	(0x1018 + 0x100 * (lane))
+#define EDP_TX_LN_VOLT_SWING_VAL_8189_FLDMASK	GENMASK(2, 1)
+#define EDP_TX_LN_VOLT_SWING_EN_8189		BIT(0)
+#define EDP_TX_LN_PRE_EMPH_VAL_8189_FLDMASK	GENMASK(6, 5)
+#define EDP_TX_LN_PRE_EMPH_EN_8189		BIT(4)
+#define RG_DSI_DEM_EN			0x500
+#define DSI_DE_EMPHASIS_ENABLE	BIT(1)
+#endif /*_MTK_DP_REG_H_*/
\ No newline at end of file
diff --git a/drivers/gpu/drm/mediatek/mtk_drm_ddp_comp.c b/drivers/gpu/drm/mediatek/mtk_drm_ddp_comp.c
index b11af62abb18..8d22ede06221 100644
--- a/drivers/gpu/drm/mediatek/mtk_drm_ddp_comp.c
+++ b/drivers/gpu/drm/mediatek/mtk_drm_ddp_comp.c
@@ -166,6 +166,11 @@ static void mtk_dither_config(struct device *dev, unsigned int w,
 			      DITHER_ENGINE_EN, cmdq_pkt);
 }
 
+static const struct mtk_ddp_comp_funcs ddp_dvo = {
+	.start = mtk_dvo_start,
+	.stop = mtk_dvo_stop,
+};
+
 static void mtk_dither_start(struct device *dev)
 {
 	struct mtk_ddp_comp_dev *priv = dev_get_drvdata(dev);
@@ -442,6 +447,7 @@ static const char * const mtk_ddp_comp_stem[MTK_DDP_COMP_TYPE_MAX] = {
 	[MTK_DP_INTF] = "dp-intf",
 	[MTK_DPI] = "dpi",
 	[MTK_DSI] = "dsi",
+	[MTK_DVO] = "dvo",
 };
 
 struct mtk_ddp_comp_match {
@@ -471,6 +477,7 @@ static const struct mtk_ddp_comp_match mtk_ddp_matches[DDP_COMPONENT_DRM_ID_MAX]
 	[DDP_COMPONENT_DSI1]		= { MTK_DSI,			1, &ddp_dsi },
 	[DDP_COMPONENT_DSI2]		= { MTK_DSI,			2, &ddp_dsi },
 	[DDP_COMPONENT_DSI3]		= { MTK_DSI,			3, &ddp_dsi },
+	[DDP_COMPONENT_DVO0]		= { MTK_DVO,			0, &ddp_dvo },
 	[DDP_COMPONENT_GAMMA0]		= { MTK_DISP_GAMMA,		0, &ddp_gamma },
 	[DDP_COMPONENT_GAMMA1]		= { MTK_DISP_GAMMA,		1, &ddp_gamma },
 	[DDP_COMPONENT_MERGE0]		= { MTK_DISP_MERGE,		0, &ddp_merge },
@@ -650,7 +657,8 @@ int mtk_ddp_comp_init(struct device_node *node, struct mtk_ddp_comp *comp,
 	    type == MTK_DISP_RDMA ||
 	    type == MTK_DPI ||
 	    type == MTK_DP_INTF ||
-	    type == MTK_DSI)
+	    type == MTK_DSI ||
+	    type == MTK_DVO)
 		return 0;
 
 	priv = devm_kzalloc(comp->dev, sizeof(*priv), GFP_KERNEL);
diff --git a/drivers/gpu/drm/mediatek/mtk_drm_drv.c b/drivers/gpu/drm/mediatek/mtk_drm_drv.c
index 7063912b3fcb..78cd7dc60469 100644
--- a/drivers/gpu/drm/mediatek/mtk_drm_drv.c
+++ b/drivers/gpu/drm/mediatek/mtk_drm_drv.c
@@ -381,9 +381,17 @@ static enum mtk_ddp_comp_id mt8189_mtk_ddp_ext[] = {
 	DDP_COMPONENT_DVO0,
 };
 
+static const unsigned int mt8189_mtk_ddp_ext_routes_0[] = {
+	DDP_COMPONENT_DVO0
+};
+
+static const unsigned int mt8189_mtk_ddp_ext_routes_1[] = {
+	DDP_COMPONENT_DSI0
+};
+
 static const struct mtk_drm_route mt8189_mtk_ddp_ext_routes[] = {
-	{1, DDP_COMPONENT_DVO0},
-	{1, DDP_COMPONENT_DP_INTF0},
+	{1, 0, ARRAY_SIZE(mt8189_mtk_ddp_ext_routes_0), mt8189_mtk_ddp_ext_routes_0},
+	{1, 0, ARRAY_SIZE(mt8189_mtk_ddp_ext_routes_1), mt8189_mtk_ddp_ext_routes_1}
 };
 
 static const struct mtk_mmsys_driver_data mt2701_mmsys_driver_data = {
@@ -503,6 +511,8 @@ static const struct mtk_mmsys_driver_data mt8365_mmsys_driver_data = {
 static const struct mtk_mmsys_driver_data mt8189_mmsys_driver_data = {
 	.main_path = mt8189_mtk_ddp_main,
 	.main_len = ARRAY_SIZE(mt8189_mtk_ddp_main),
+	.ext_path = mt8189_mtk_ddp_ext,
+	.ext_len = ARRAY_SIZE(mt8189_mtk_ddp_ext),
 	.mmsys_dev_num = 1,
 };
 
@@ -1048,6 +1058,10 @@ static const struct of_device_id mtk_ddp_comp_dt_ids[] = {
 	  .data = (void *)MTK_DPI },
 	{ .compatible = "mediatek,mt8188-dp-intf",
 	  .data = (void *)MTK_DP_INTF },
+	{ .compatible = "mediatek,mt8189-dp-dvo",
+	  .data = (void *)MTK_DVO },
+	{ .compatible = "mediatek,mt8189-edp-dvo",
+	  .data = (void *)MTK_DVO },
 	{ .compatible = "mediatek,mt8192-dpi",
 	  .data = (void *)MTK_DPI },
 	{ .compatible = "mediatek,mt8195-dpi",
@@ -1172,8 +1186,8 @@ static int mtk_drm_probe(struct platform_device *pdev)
 		    comp_type == MTK_DISP_RDMA ||
 		    comp_type == MTK_DP_INTF ||
 		    comp_type == MTK_DPI ||
-		    comp_type == MTK_DVO ||
-		    comp_type == MTK_DSI) {
+		    comp_type == MTK_DSI ||
+		    comp_type == MTK_DVO) {
 			if (!drm_comp_add_list[comp_id]) {
 				dev_info(dev, "Adding component match for %pOF\n",
 					 node);
@@ -1285,6 +1299,7 @@ static struct platform_driver * const mtk_drm_drivers[] = {
 	&mtk_disp_ovl_driver,
 	&mtk_disp_rdma_driver,
 	&mtk_dpi_driver,
+	&mtk_dvo_driver,
 	&mtk_drm_platform_driver,
 	&mtk_dsi_driver,
 	&mtk_ethdr_driver,
diff --git a/drivers/gpu/drm/mediatek/mtk_drm_drv.h b/drivers/gpu/drm/mediatek/mtk_drm_drv.h
index b2ddc1f407b4..51438b939ea7 100644
--- a/drivers/gpu/drm/mediatek/mtk_drm_drv.h
+++ b/drivers/gpu/drm/mediatek/mtk_drm_drv.h
@@ -86,4 +86,5 @@ extern struct platform_driver mtk_mdp_rdma_driver;
 extern struct platform_driver mtk_dpintf_driver;
 extern struct platform_driver mtk_dp_tx_driver;
 extern struct platform_driver mtk_disp_padding_driver;
+extern struct platform_driver mtk_dvo_driver;
 #endif /* MTK_DRM_DRV_H */
diff --git a/drivers/gpu/drm/mediatek/mtk_dvo.c b/drivers/gpu/drm/mediatek/mtk_dvo.c
new file mode 100644
index 000000000000..c6677e097381
--- /dev/null
+++ b/drivers/gpu/drm/mediatek/mtk_dvo.c
@@ -0,0 +1,885 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) 2014 MediaTek Inc.
+ * Author: Jie Qiu <jie.qiu@mediatek.com>
+ */
+
+#include <linux/clk.h>
+#include <linux/component.h>
+#include <linux/interrupt.h>
+#include <linux/kernel.h>
+#include <linux/media-bus-format.h>
+#include <linux/of.h>
+#include <linux/of_graph.h>
+#include <linux/pinctrl/consumer.h>
+#include <linux/platform_device.h>
+#include <linux/soc/mediatek/mtk-mmsys.h>
+#include <linux/types.h>
+
+#include <video/videomode.h>
+
+#include <drm/drm_atomic_helper.h>
+#include <drm/drm_bridge.h>
+#include <drm/drm_bridge_connector.h>
+#include <drm/drm_crtc.h>
+#include <drm/drm_edid.h>
+#include <drm/drm_of.h>
+#include <drm/drm_simple_kms_helper.h>
+
+#include "mtk_dvo_regs.h"
+#include "mtk_drm_drv.h"
+#include "mtk_drm_crtc.h"
+#include "mtk_drm_ddp_comp.h"
+
+enum mtk_dvo_out_bit_num {
+	MTK_DVO_OUT_BIT_NUM_8BITS,
+	MTK_DVO_OUT_BIT_NUM_10BITS,
+	MTK_DVO_OUT_BIT_NUM_12BITS,
+	MTK_DVO_OUT_BIT_NUM_16BITS
+};
+
+enum mtk_dvo_out_yc_map {
+	MTK_DVO_OUT_YC_MAP_RGB,
+	MTK_DVO_OUT_YC_MAP_CYCY,
+	MTK_DVO_OUT_YC_MAP_YCYC,
+	MTK_DVO_OUT_YC_MAP_CY,
+	MTK_DVO_OUT_YC_MAP_YC
+};
+
+enum mtk_dvo_out_channel_swap {
+	MTK_DVO_OUT_CHANNEL_SWAP_RGB,
+	MTK_DVO_OUT_CHANNEL_SWAP_GBR,
+	MTK_DVO_OUT_CHANNEL_SWAP_BRG,
+	MTK_DVO_OUT_CHANNEL_SWAP_RBG,
+	MTK_DVO_OUT_CHANNEL_SWAP_GRB,
+	MTK_DVO_OUT_CHANNEL_SWAP_BGR
+};
+
+enum mtk_dvo_out_color_format {
+	MTK_DVO_COLOR_FORMAT_RGB,
+	MTK_DVO_COLOR_FORMAT_YCBCR_422
+};
+
+enum TVDPLL_CLK {
+	TVDPLL_PLL = 0,
+	TVDPLL_D2 = 2,
+	TVDPLL_D4 = 4,
+	TVDPLL_D8 = 8,
+	TVDPLL_D16 = 16,
+};
+
+enum mtk_dvo_golden_setting_level {
+	MTK_DVO_FHD_60FPS_1920 = 0,
+	MTK_DVO_FHD_60FPS_2180,
+	MTK_DVO_FHD_60FPS_2400,
+	MTK_DVO_FHD_60FPS_2520,
+	MTK_DVO_FHD_90FPS,
+	MTK_DVO_FHD_120FPS,
+	MTK_DVO_WQHD_60FPS,
+	MTK_DVO_WQHD_120FPS,
+	MTK_DVO_8K_30FPS,
+	MTK_DVO_GSL_MAX,
+};
+
+struct mtk_dvo_gs_info {
+	u32 dvo_buf_sodi_high;
+	u32 dvo_buf_sodi_low;
+};
+
+struct mtk_dvo {
+	struct mtk_ddp_comp ddp_comp;
+	struct drm_encoder encoder;
+	struct drm_bridge bridge;
+	struct drm_bridge *next_bridge;
+	struct drm_connector *connector;
+	void __iomem *regs;
+	struct device *dev;
+	struct device *mmsys_dev;
+	struct clk *engine_clk;
+	struct clk *pixel_clk;
+	struct clk *tvd_clk;
+	struct clk *dvo_clk;
+	struct clk *hf_fdvo_clk;
+	struct clk *pclk_src[5];
+	int irq;
+	struct drm_display_mode mode;
+	const struct mtk_dvo_conf *conf;
+	enum mtk_dvo_out_color_format color_format;
+	enum mtk_dvo_out_yc_map yc_map;
+	enum mtk_dvo_out_bit_num bit_num;
+	enum mtk_dvo_out_channel_swap channel_swap;
+	struct pinctrl *pinctrl;
+	struct pinctrl_state *pins_gpio;
+	struct pinctrl_state *pins_dvo;
+	u32 output_fmt;
+	int refcount;
+	enum mtk_dvo_golden_setting_level gs_level;
+};
+
+static inline struct mtk_dvo *bridge_to_dvo(struct drm_bridge *b)
+{
+	return container_of(b, struct mtk_dvo, bridge);
+}
+
+enum mtk_dvo_polarity {
+	MTK_DVO_POLARITY_RISING,
+	MTK_DVO_POLARITY_FALLING,
+};
+
+struct mtk_dvo_polarities {
+	enum mtk_dvo_polarity de_pol;
+	enum mtk_dvo_polarity ck_pol;
+	enum mtk_dvo_polarity hsync_pol;
+	enum mtk_dvo_polarity vsync_pol;
+};
+
+struct mtk_dvo_sync_param {
+	u32 sync_width;
+	u32 front_porch;
+	u32 back_porch;
+	bool shift_half_line;
+};
+
+struct mtk_dvo_yc_limit {
+	u16 y_top;
+	u16 y_bottom;
+	u16 c_top;
+	u16 c_bottom;
+};
+
+/**
+ * struct mtk_dvo_conf - Configuration of mediatek dvo.
+ * @cal_factor: Callback function to calculate factor value.
+ * @reg_h_fre_con: Register address of frequency control.
+ * @max_clock_khz: Max clock frequency supported for this SoCs in khz units.
+ * @edge_sel_en: Enable of edge selection.
+ * @output_fmts: Array of supported output formats.
+ * @num_output_fmts: Quantity of supported output formats.
+ * @is_ck_de_pol: Support CK/DE polarity.
+ * @swap_input_support: Support input swap function.
+ * @support_direct_pin: IP supports direct connection to dvo panels.
+ * @input_2pixel: Input pixel of dp_intf is 2 pixel per round, so enable this
+ *		  config to enable this feature.
+ * @dimension_mask: Mask used for HWIDTH, HPORCH, VSYNC_WIDTH and VSYNC_PORCH
+ *		    (no shift).
+ * @hvsize_mask: Mask of HSIZE and VSIZE mask (no shift).
+ * @channel_swap_shift: Shift value of channel swap.
+ * @yuv422_en_bit: Enable bit of yuv422.
+ * @csc_enable_bit: Enable bit of CSC.
+ * @pixels_per_iter: Quantity of transferred pixels per iteration.
+ * @edge_cfg_in_mmsys: If the edge configuration for DVO's output needs to be set in MMSYS.
+ */
+struct mtk_dvo_conf {
+	unsigned int (*cal_factor)(int clock);
+	u32 reg_h_fre_con;
+	u32 max_clock_khz;
+	bool edge_sel_en;
+	const u32 *output_fmts;
+	u32 num_output_fmts;
+	bool is_ck_de_pol;
+	bool swap_input_support;
+	bool support_direct_pin;
+	bool input_2pixel;
+	u32 out_np_sel;
+	u32 dimension_mask;
+	u32 hvsize_mask;
+	u32 channel_swap_shift;
+	u32 yuv422_en_bit;
+	u32 csc_enable_bit;
+	u32 pixels_per_iter;
+	bool edge_cfg_in_mmsys;
+	bool has_commit;
+	bool is_dp;
+};
+
+static struct mtk_dvo_gs_info mtk_dvo_gs[MTK_DVO_GSL_MAX] = {
+	[MTK_DVO_FHD_60FPS_1920] = {6880, 511},
+	[MTK_DVO_8K_30FPS] = {5255, 3899},
+};
+
+static void mtk_dvo_mask(struct mtk_dvo *dvo, u32 offset, u32 val, u32 mask)
+{
+	u32 tmp = readl(dvo->regs + offset) & ~mask;
+
+	tmp |= (val & mask);
+	writel(tmp, dvo->regs + offset);
+}
+
+static void mtk_dvo_sw_reset(struct mtk_dvo *dvo, bool reset)
+{
+	mtk_dvo_mask(dvo, DVO_RET, reset ? SWRST : 0, SWRST);
+}
+
+static void mtk_dvo_enable(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_EN, EN, EN);
+}
+
+static void mtk_dvo_disable(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_EN, 0, EN);
+}
+
+static void mtk_dvo_irq_enable(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_INTEN, INT_VDE_END_EN, INT_VDE_END_EN);
+}
+
+static void mtk_dvo_info_queue_start(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_TGEN_INFOQ_LATENCY, 0,
+		     INFOQ_START_LATENCY_MASK | INFOQ_END_LATENCY_MASK);
+}
+
+static void mtk_dvo_buffer_ctrl(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_BUF_CON0, DISP_BUF_EN, DISP_BUF_EN);
+	mtk_dvo_mask(dvo, DVO_BUF_CON0, FIFO_UNDERFLOW_DONE_BLOCK, FIFO_UNDERFLOW_DONE_BLOCK);
+}
+
+static void mtk_dvo_trailing_blank_setting(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_TGEN_V_LAST_TRAILING_BLANK, 0x20, V_LAST_TRAILING_BLANK_MASK);
+	mtk_dvo_mask(dvo, DVO_TGEN_OUTPUT_DELAY_LINE, 0x20, EXT_TG_DLY_LINE_MASK);
+}
+
+static void mtk_dvo_get_gs_level(struct mtk_dvo *dvo)
+{
+	struct drm_display_mode *mode = &dvo->mode;
+	enum mtk_dvo_golden_setting_level *gsl = &dvo->gs_level;
+
+	if (mode->hdisplay == 1920 && mode->vdisplay == 1080)
+		*gsl = MTK_DVO_FHD_60FPS_1920;
+	else
+		*gsl = MTK_DVO_8K_30FPS;
+}
+
+static void mtk_dvo_golden_setting(struct mtk_dvo *dvo)
+{
+	struct mtk_dvo_gs_info *gs_info = NULL;
+
+	if (dvo->gs_level >= MTK_DVO_GSL_MAX) {
+		dev_info(dvo->dev,"%s invalid gs_level %d\n",
+			 __func__, dvo->gs_level);
+		return;
+	}
+
+	gs_info = &mtk_dvo_gs[dvo->gs_level];
+
+	mtk_dvo_mask(dvo, DVO_BUF_SODI_HIGHT, gs_info->dvo_buf_sodi_high, GENMASK(31,0));
+	mtk_dvo_mask(dvo, DVO_BUF_SODI_LOW, gs_info->dvo_buf_sodi_low, GENMASK(31,0));
+}
+
+static void mtk_dvo_shadow_ctrl(struct mtk_dvo *dvo)
+{
+	mtk_dvo_mask(dvo, DVO_SHADOW_CTRL, 0, BYPASS_SHADOW);
+	mtk_dvo_mask(dvo, DVO_SHADOW_CTRL, FORCE_COMMIT, FORCE_COMMIT);
+}
+
+static void mtk_dvo_config_hsync(struct mtk_dvo *dvo,
+				 struct mtk_dvo_sync_param *sync)
+{
+	mtk_dvo_mask(dvo, DVO_TGEN_H0, sync->sync_width << HSYNC,
+		     dvo->conf->dimension_mask << HSYNC);
+	mtk_dvo_mask(dvo, DVO_TGEN_H0, sync->front_porch << HFP,
+		     dvo->conf->dimension_mask << HFP);
+	mtk_dvo_mask(dvo, DVO_TGEN_H1, (sync->back_porch + sync->sync_width) << HSYNC2ACT,
+		     dvo->conf->dimension_mask << HSYNC2ACT);
+}
+
+static void mtk_dvo_config_vsync(struct mtk_dvo *dvo,
+				 struct mtk_dvo_sync_param *sync,
+				 u32 width_addr, u32 porch_addr)
+{
+	mtk_dvo_mask(dvo, width_addr,
+		     sync->sync_width << VSYNC,
+		     dvo->conf->dimension_mask << VSYNC);
+	mtk_dvo_mask(dvo, width_addr,
+		     sync->front_porch << VFP,
+		     dvo->conf->dimension_mask << VFP);
+	mtk_dvo_mask(dvo, porch_addr,
+		     (sync->back_porch + sync->sync_width) << VSYNC2ACT,
+	     dvo->conf->dimension_mask << VSYNC2ACT);
+}
+
+static void mtk_dvo_config_vsync_lodd(struct mtk_dvo *dvo,
+				      struct mtk_dvo_sync_param *sync)
+{
+	mtk_dvo_config_vsync(dvo, sync, DVO_TGEN_V0, DVO_TGEN_V1);
+}
+
+static void mtk_dvo_config_interface(struct mtk_dvo *dvo, bool inter)
+{
+	mtk_dvo_mask(dvo, DVO_CON, inter ? INTL_EN : 0, INTL_EN);
+}
+
+static void mtk_dvo_config_fb_size(struct mtk_dvo *dvo, u32 width, u32 height)
+{
+	mtk_dvo_mask(dvo, DVO_SRC_SIZE, width << SRC_HSIZE,
+		     dvo->conf->hvsize_mask << SRC_HSIZE);
+	mtk_dvo_mask(dvo, DVO_SRC_SIZE, height << SRC_VSIZE,
+		     dvo->conf->hvsize_mask << SRC_VSIZE);
+
+	mtk_dvo_mask(dvo, DVO_PIC_SIZE, width << PIC_HSIZE,
+		     dvo->conf->hvsize_mask << PIC_HSIZE);
+	mtk_dvo_mask(dvo, DVO_PIC_SIZE, height << PIC_VSIZE,
+		     dvo->conf->hvsize_mask << PIC_VSIZE);
+
+	mtk_dvo_mask(dvo, DVO_TGEN_H1, DIV_ROUND_UP(width, dvo->conf->pixels_per_iter) << HACT,
+		     dvo->conf->hvsize_mask << HACT);
+
+	mtk_dvo_mask(dvo, DVO_TGEN_V1, height << VACT,
+		     dvo->conf->hvsize_mask << VACT);
+}
+
+static void mtk_dvo_power_off(struct mtk_dvo *dvo)
+{
+	if (WARN_ON(dvo->refcount == 0))
+		return;
+
+	if (--dvo->refcount != 0)
+		return;
+
+	mtk_dvo_disable(dvo);
+
+	clk_disable_unprepare(dvo->engine_clk);
+	clk_disable_unprepare(dvo->tvd_clk);
+	clk_disable_unprepare(dvo->hf_fdvo_clk);
+}
+
+static int mtk_dvo_power_on(struct mtk_dvo *dvo)
+{
+	int ret;
+
+	if (++dvo->refcount != 1)
+		return 0;
+
+	ret = clk_prepare_enable(dvo->hf_fdvo_clk);
+	if (ret) {
+		dev_err(dvo->dev, "Failed to enable hf_fdvo_clk clock: %d\n", ret);
+		goto err_refcount;
+	}
+
+	ret = clk_prepare_enable(dvo->tvd_clk);
+	if (ret) {
+		dev_err(dvo->dev, "Failed to enable tvd_clk clock: %d\n", ret);
+		goto err_refcount;
+	}
+
+	ret = clk_prepare_enable(dvo->engine_clk);
+	if (ret) {
+		dev_err(dvo->dev, "Failed to enable engine clock: %d\n", ret);
+		goto err_refcount;
+	}
+
+	ret = clk_prepare_enable(dvo->dvo_clk);
+	if (ret) {
+		dev_err(dvo->dev, "Failed to enable engine clock: %d\n", ret);
+		goto err_refcount;
+	}
+
+	ret = clk_prepare_enable(dvo->pixel_clk);
+	if (ret) {
+		dev_err(dvo->dev, "Failed to enable engine clock: %d\n", ret);
+		goto err_refcount;
+	}
+
+	return 0;
+
+err_refcount:
+	dvo->refcount--;
+	return ret;
+}
+
+static int mtk_dvo_set_display_mode(struct mtk_dvo *dvo,
+				    struct drm_display_mode *mode)
+{
+	struct mtk_dvo_sync_param hsync;
+	struct mtk_dvo_sync_param vsync_lodd = { 0 };
+	struct videomode vm = { 0 };
+	unsigned long pll_rate;
+	unsigned int factor;
+
+	pr_info("[eDPTX] %s+\n", __func__);
+
+	/* let pll_rate can fix the valid range of tvdpll (1G~2GHz) */
+	factor = dvo->conf->cal_factor(mode->clock);
+	drm_display_mode_to_videomode(mode, &vm);
+	pll_rate = vm.pixelclock * factor;
+
+	dev_dbg(dvo->dev, "Want PLL %lu Hz, pixel clock %lu Hz\n",
+			   pll_rate, vm.pixelclock);
+
+	clk_set_rate(dvo->tvd_clk, pll_rate);
+	pll_rate = clk_get_rate(dvo->tvd_clk);
+
+	/*
+	 * Depending on the IP version, we may output a different amount of
+	 * pixels for each iteration: divide the clock by this number and
+	 * adjust the display porches accordingly.
+	 */
+	vm.pixelclock = pll_rate / factor;
+	vm.pixelclock /= dvo->conf->pixels_per_iter;
+	clk_set_rate(dvo->pixel_clk, vm.pixelclock);
+
+	vm.pixelclock = clk_get_rate(dvo->pixel_clk);
+
+	dev_dbg(dvo->dev, "Got  PLL %lu Hz, pixel clock %lu Hz\n",
+			   pll_rate, vm.pixelclock);
+	/*
+	 * Depending on the IP version, we may output a different amount of
+	 * pixels for each iteration: divide the clock by this number and
+	 * adjust the display porches accordingly.
+	 */
+	hsync.sync_width = vm.hsync_len / dvo->conf->pixels_per_iter;
+	hsync.back_porch = vm.hback_porch / dvo->conf->pixels_per_iter;
+	hsync.front_porch = vm.hfront_porch / dvo->conf->pixels_per_iter;
+
+	hsync.shift_half_line = false;
+	vsync_lodd.sync_width = vm.vsync_len;
+	vsync_lodd.back_porch = vm.vback_porch;
+	vsync_lodd.front_porch = vm.vfront_porch;
+	vsync_lodd.shift_half_line = false;
+
+	mtk_dvo_sw_reset(dvo, true);
+	mtk_dvo_irq_enable(dvo);
+
+	mtk_dvo_config_hsync(dvo, &hsync);
+	mtk_dvo_config_vsync_lodd(dvo, &vsync_lodd);
+
+	mtk_dvo_config_interface(dvo, !!(vm.flags & DISPLAY_FLAGS_INTERLACED));
+
+	if (vm.flags & DISPLAY_FLAGS_INTERLACED)
+		mtk_dvo_config_fb_size(dvo, vm.hactive, vm.vactive >> 1);
+	else
+		mtk_dvo_config_fb_size(dvo, vm.hactive, vm.vactive);
+
+	mtk_dvo_info_queue_start(dvo);
+	mtk_dvo_buffer_ctrl(dvo);
+	if (!dvo->conf->is_dp) {
+		mtk_dvo_trailing_blank_setting(dvo);
+		mtk_dvo_get_gs_level(dvo);
+		mtk_dvo_golden_setting(dvo);
+	}
+
+	if (dvo->conf->pixels_per_iter)
+		mtk_dvo_shadow_ctrl(dvo);
+	if (dvo->conf->out_np_sel) {
+		mtk_dvo_mask(dvo, DVO_OUTPUT_SET, dvo->conf->out_np_sel,
+			     OUT_NP_SEL);
+	}
+	mtk_dvo_sw_reset(dvo, false);
+
+	return 0;
+}
+
+static u32 *mtk_dvo_bridge_atomic_get_output_bus_fmts(struct drm_bridge *bridge,
+						      struct drm_bridge_state *bridge_state,
+						      struct drm_crtc_state *crtc_state,
+						      struct drm_connector_state *conn_state,
+						      unsigned int *num_output_fmts)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+	u32 *output_fmts;
+
+	*num_output_fmts = 0;
+
+	if (!dvo->conf->output_fmts) {
+		dev_err(dvo->dev, "output_fmts should not be null\n");
+		return NULL;
+	}
+
+	output_fmts = kcalloc(dvo->conf->num_output_fmts, sizeof(*output_fmts),
+			      GFP_KERNEL);
+	if (!output_fmts)
+		return NULL;
+
+	*num_output_fmts = dvo->conf->num_output_fmts;
+
+	memcpy(output_fmts, dvo->conf->output_fmts,
+	       sizeof(*output_fmts) * dvo->conf->num_output_fmts);
+
+	return output_fmts;
+}
+
+static u32 *mtk_dvo_bridge_atomic_get_input_bus_fmts(struct drm_bridge *bridge,
+						     struct drm_bridge_state *bridge_state,
+						     struct drm_crtc_state *crtc_state,
+						     struct drm_connector_state *conn_state,
+						     u32 output_fmt,
+						     unsigned int *num_input_fmts)
+{
+	u32 *input_fmts;
+
+	*num_input_fmts = 0;
+
+	input_fmts = kcalloc(1, sizeof(*input_fmts),
+			     GFP_KERNEL);
+	if (!input_fmts)
+		return NULL;
+
+	*num_input_fmts = 1;
+	input_fmts[0] = MEDIA_BUS_FMT_RGB888_1X24;
+
+	return input_fmts;
+}
+
+static int mtk_dvo_bridge_atomic_check(struct drm_bridge *bridge,
+				       struct drm_bridge_state *bridge_state,
+				       struct drm_crtc_state *crtc_state,
+				       struct drm_connector_state *conn_state)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+	unsigned int out_bus_format;
+
+	out_bus_format = bridge_state->output_bus_cfg.format;
+
+	if (out_bus_format == MEDIA_BUS_FMT_FIXED)
+		if (dvo->conf->num_output_fmts)
+			out_bus_format = dvo->conf->output_fmts[0];
+
+	dev_dbg(dvo->dev, "input format 0x%04x, output format 0x%04x\n",
+		bridge_state->input_bus_cfg.format,
+		bridge_state->output_bus_cfg.format);
+
+	dvo->output_fmt = out_bus_format;
+	dvo->bit_num = MTK_DVO_OUT_BIT_NUM_8BITS;
+	dvo->channel_swap = MTK_DVO_OUT_CHANNEL_SWAP_RGB;
+	dvo->yc_map = MTK_DVO_OUT_YC_MAP_RGB;
+	if (out_bus_format == MEDIA_BUS_FMT_YUYV8_1X16)
+		dvo->color_format = MTK_DVO_COLOR_FORMAT_YCBCR_422;
+	else
+		dvo->color_format = MTK_DVO_COLOR_FORMAT_RGB;
+
+	return 0;
+}
+
+static int mtk_dvo_bridge_attach(struct drm_bridge *bridge,
+				 enum drm_bridge_attach_flags flags)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+	int ret = 0;
+	dev_info(dvo->dev, "[eDPTX] %s %d\n", __func__, __LINE__);
+
+	ret = drm_bridge_attach(bridge->encoder, dvo->next_bridge,
+				&dvo->bridge, 1);
+	if (ret)
+		dev_err(dvo->dev, " Failed to call attach ret = %d\n", ret);
+
+	return ret;
+}
+
+static void mtk_dvo_bridge_mode_set(struct drm_bridge *bridge,
+		const struct drm_display_mode *mode,
+		const struct drm_display_mode *adjusted_mode)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+
+	drm_mode_copy(&dvo->mode, adjusted_mode);
+}
+
+static void mtk_dvo_bridge_disable(struct drm_bridge *bridge)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+
+	mtk_dvo_power_off(dvo);
+
+	if (dvo->pinctrl && dvo->pins_gpio)
+		pinctrl_select_state(dvo->pinctrl, dvo->pins_gpio);
+}
+
+static void mtk_dvo_bridge_enable(struct drm_bridge *bridge)
+{
+	struct mtk_dvo *dvo = bridge_to_dvo(bridge);
+
+	if (dvo->pinctrl && dvo->pins_dvo)
+		pinctrl_select_state(dvo->pinctrl, dvo->pins_dvo);
+
+	mtk_dvo_power_on(dvo);
+	mtk_dvo_set_display_mode(dvo, &dvo->mode);
+	mtk_dvo_enable(dvo);
+	/*
+	 * mtk_dvo_mask(dvo, DVO_PATTERN_CTRL, PRE_PAT_EN, PRE_PAT_EN);
+	 * mtk_dvo_mask(dvo, DVO_PATTERN_CTRL, COLOR_BAR, PRE_PAT_SEL_MASK);
+	 * mtk_dvo_mask(dvo, DVO_PATTERN_CTRL, PRE_PAT_FORCE_ON, PRE_PAT_FORCE_ON);
+	 * mtk_dvo_mask(dvo, DVO_PATTERN_COLOR, PAT_G, PAT_G);
+	 */
+}
+
+static enum drm_mode_status
+mtk_dvo_bridge_mode_valid(struct drm_bridge *bridge,
+			  const struct drm_display_info *info,
+			  const struct drm_display_mode *mode)
+{
+	return MODE_OK;
+}
+
+static const struct drm_bridge_funcs mtk_dvo_bridge_funcs = {
+	.attach = mtk_dvo_bridge_attach,
+	.mode_set = mtk_dvo_bridge_mode_set,
+	.mode_valid = mtk_dvo_bridge_mode_valid,
+	.disable = mtk_dvo_bridge_disable,
+	.enable = mtk_dvo_bridge_enable,
+	.atomic_check = mtk_dvo_bridge_atomic_check,
+	.atomic_get_output_bus_fmts = mtk_dvo_bridge_atomic_get_output_bus_fmts,
+	.atomic_get_input_bus_fmts = mtk_dvo_bridge_atomic_get_input_bus_fmts,
+	.atomic_duplicate_state = drm_atomic_helper_bridge_duplicate_state,
+	.atomic_destroy_state = drm_atomic_helper_bridge_destroy_state,
+	.atomic_reset = drm_atomic_helper_bridge_reset,
+};
+
+void mtk_dvo_start(struct device *dev)
+{
+	struct mtk_dvo *dvo = dev_get_drvdata(dev);
+
+	mtk_dvo_power_on(dvo);
+}
+
+void mtk_dvo_stop(struct device *dev)
+{
+	struct mtk_dvo *dvo = dev_get_drvdata(dev);
+
+	mtk_dvo_power_off(dvo);
+}
+
+static int mtk_dvo_bind(struct device *dev, struct device *master, void *data)
+{
+	struct mtk_dvo *dvo = dev_get_drvdata(dev);
+	struct drm_device *drm_dev = data;
+	struct mtk_drm_private *priv = drm_dev->dev_private;
+	int ret;
+	dev_info(dev, "[eDPTX] %s %d\n", __func__, __LINE__);
+
+	dvo->mmsys_dev = priv->mmsys_dev;
+	ret = drm_simple_encoder_init(drm_dev, &dvo->encoder,
+				      DRM_MODE_ENCODER_TMDS);
+	if (ret) {
+		dev_err(dev, "Failed to initialize decoder: %d\n", ret);
+		return ret;
+	}
+	dev_info(dev, "[eDPTX] %s %d\n", __func__, __LINE__);
+	dvo->encoder.possible_crtcs = mtk_drm_find_possible_crtc_by_comp(drm_dev, dvo->dev);
+
+	ret = drm_bridge_attach(&dvo->encoder, &dvo->bridge, NULL,
+				DRM_BRIDGE_ATTACH_NO_CONNECTOR);
+	if (ret)
+		goto err_cleanup;
+	dev_info(dev, "[eDPTX] %s %d\n", __func__, __LINE__);
+
+	dvo->connector = drm_bridge_connector_init(drm_dev, &dvo->encoder);
+	if (IS_ERR(dvo->connector)) {
+		dev_err(dev, "Unable to create bridge connector\n");
+		ret = PTR_ERR(dvo->connector);
+		goto err_cleanup;
+	}
+	drm_connector_attach_encoder(dvo->connector, &dvo->encoder);
+
+	dev_info(dev, "[eDPTX] %s-\n", __func__);
+
+	return 0;
+
+err_cleanup:
+	drm_encoder_cleanup(&dvo->encoder);
+	return ret;
+}
+
+static void mtk_dvo_unbind(struct device *dev, struct device *master,
+			   void *data)
+{
+	struct mtk_dvo *dvo = dev_get_drvdata(dev);
+
+	drm_encoder_cleanup(&dvo->encoder);
+}
+
+static const struct component_ops mtk_dvo_component_ops = {
+	.bind = mtk_dvo_bind,
+	.unbind = mtk_dvo_unbind,
+};
+
+static unsigned int mt8196_calculate_factor(int clock)
+{
+	if (clock < 70000)
+		return 4;
+	else if (clock < 200000)
+		return 2;
+	else
+		return 1;
+}
+
+static const u32 mt8196_output_fmts[] = {
+	MEDIA_BUS_FMT_RGB888_1X24,
+	MEDIA_BUS_FMT_YUYV8_1X16,
+};
+
+static const struct mtk_dvo_conf mt8189_conf = {
+	.cal_factor = mt8196_calculate_factor,
+	.out_np_sel = 0X2,
+	.reg_h_fre_con = 0xb0,
+	.max_clock_khz = 640000,
+	.output_fmts = mt8196_output_fmts,
+	.num_output_fmts = ARRAY_SIZE(mt8196_output_fmts),
+	.pixels_per_iter = 4,
+	.has_commit = true,
+	.dimension_mask = HFP_MASK,
+	.hvsize_mask = PIC_HSIZE_MASK,
+};
+
+static const struct mtk_dvo_conf mt8189_dp_conf = {
+	.cal_factor = mt8196_calculate_factor,
+	.out_np_sel = 0X2,
+	.reg_h_fre_con = 0xb0,
+	.max_clock_khz = 640000,
+	.output_fmts = mt8196_output_fmts,
+	.num_output_fmts = ARRAY_SIZE(mt8196_output_fmts),
+	.pixels_per_iter = 4,
+	.has_commit = true,
+	.dimension_mask = HFP_MASK,
+	.hvsize_mask = PIC_HSIZE_MASK,
+	.is_dp = true,
+};
+
+static int mtk_dvo_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct device_node *endpoint, *port;
+	struct mtk_dvo *dvo;
+	int ret;
+
+	dvo = devm_kzalloc(dev, sizeof(*dvo), GFP_KERNEL);
+	if (!dvo)
+		return -ENOMEM;
+
+	dvo->dev = dev;
+	dvo->conf = (struct mtk_dvo_conf *)of_device_get_match_data(dev);
+	dvo->output_fmt = MEDIA_BUS_FMT_RGB888_1X24;
+
+	dvo->pinctrl = devm_pinctrl_get(&pdev->dev);
+	if (IS_ERR(dvo->pinctrl)) {
+		dvo->pinctrl = NULL;
+		dev_dbg(&pdev->dev, "Cannot find pinctrl!\n");
+	}
+	if (dvo->pinctrl) {
+		dvo->pins_gpio = pinctrl_lookup_state(dvo->pinctrl, "sleep");
+		if (IS_ERR(dvo->pins_gpio)) {
+			dvo->pins_gpio = NULL;
+			dev_dbg(&pdev->dev, "Cannot find pinctrl idle!\n");
+		}
+		if (dvo->pins_gpio)
+			pinctrl_select_state(dvo->pinctrl, dvo->pins_gpio);
+
+		dvo->pins_dvo = pinctrl_lookup_state(dvo->pinctrl, "default");
+		if (IS_ERR(dvo->pins_dvo)) {
+			dvo->pins_dvo = NULL;
+			dev_dbg(&pdev->dev, "Cannot find pinctrl active!\n");
+		}
+	}
+	dvo->regs = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(dvo->regs))
+		return dev_err_probe(dev, PTR_ERR(dvo->regs),
+				     "Failed to ioremap mem resource\n");
+
+	dvo->engine_clk = devm_clk_get(dev, "engine");
+	if (IS_ERR(dvo->engine_clk))
+		return dev_err_probe(dev, PTR_ERR(dvo->engine_clk),
+				     "Failed to get engine clock\n");
+
+	dvo->pixel_clk = devm_clk_get(dev, "pixel");
+	if (IS_ERR(dvo->pixel_clk))
+		return dev_err_probe(dev, PTR_ERR(dvo->pixel_clk),
+				     "Failed to get pixel clock\n");
+
+
+	dvo->tvd_clk = devm_clk_get(dev, "pll");
+	if (IS_ERR(dvo->tvd_clk))
+		return dev_err_probe(dev, PTR_ERR(dvo->tvd_clk),
+				     "Failed to get tvdpll clock\n");
+
+	dvo->hf_fdvo_clk = devm_clk_get(dev, "hf_fdvo_clk");
+	if (IS_ERR(dvo->hf_fdvo_clk))
+		return dev_err_probe(dev, PTR_ERR(dvo->hf_fdvo_clk),
+				     "Failed to get hf_fdvo_clk clock\n");
+
+	dvo->dvo_clk = devm_clk_get_optional(dev, "dvo_clk");
+	if (IS_ERR(dvo->dvo_clk))
+		return dev_err_probe(dev, PTR_ERR(dvo->dvo_clk),
+				     "Failed to get dvo_clk clock\n");
+
+	//TODO: TVDPLL_D* are parent clock divisors which needs to be
+	//set dynamically in the clock framework and not acquired/changed here
+	dvo->pclk_src[1] = devm_clk_get_optional(dev, "TVDPLL_D2");
+	dvo->pclk_src[2] = devm_clk_get_optional(dev, "TVDPLL_D4");
+	dvo->pclk_src[3] = devm_clk_get_optional(dev, "TVDPLL_D8");
+	dvo->pclk_src[4] = devm_clk_get_optional(dev, "TVDPLL_D16");
+
+	dvo->irq = platform_get_irq(pdev, 0);
+	if (dvo->irq < 0)
+		return dvo->irq;
+
+	platform_set_drvdata(pdev, dvo);
+
+	if (!dvo->conf->is_dp) {
+		dvo->next_bridge = devm_drm_of_get_bridge(dev, dev->of_node, 0, 0);
+		if (IS_ERR(dvo->next_bridge)) {
+			return  dev_err_probe(dev, PTR_ERR(dvo->next_bridge),
+					"Failed to get bridge\n");
+		}
+	} else {
+		/* To support MST, the DP bridge (dp-tx) driver registers one bridge per input port. */
+		endpoint = of_graph_get_endpoint_by_regs(dev->of_node, 0, 0);
+		if (!endpoint) {
+			dev_err(dev, "fail to find endpoint!\n");
+			return -ENODEV;
+		}
+
+		port = of_graph_get_remote_port(endpoint);
+		of_node_put(endpoint);
+		if (!port) {
+			dev_err(dev, "fail to find remote port!\n");
+			return -ENODEV;
+		}
+
+		dvo->next_bridge = of_drm_find_bridge(port);
+		of_node_put(port);
+		if (!dvo->next_bridge) {
+			dev_err(dev, "fail to find next bridge\n");
+			return -EPROBE_DEFER;
+		}
+	}
+
+	dev_info(dev, "Found bridge node: %pOF\n", dvo->next_bridge->of_node);
+
+	dvo->bridge.funcs = &mtk_dvo_bridge_funcs;
+	dvo->bridge.of_node = dev->of_node;
+	dvo->bridge.type = DRM_MODE_CONNECTOR_DPI;
+
+	ret = devm_drm_bridge_add(dev, &dvo->bridge);
+	if (ret)
+		return ret;
+
+	ret = component_add(dev, &mtk_dvo_component_ops);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to add component.\n");
+
+	return 0;
+}
+
+static void mtk_dvo_remove(struct platform_device *pdev)
+{
+	component_del(&pdev->dev, &mtk_dvo_component_ops);
+}
+
+static const struct of_device_id mtk_dvo_of_ids[] = {
+	{ .compatible = "mediatek,mt8189-edp-dvo", .data = &mt8189_conf },
+	{ .compatible = "mediatek,mt8189-dp-dvo", .data = &mt8189_dp_conf },
+	{ /* sentinel */ },
+};
+MODULE_DEVICE_TABLE(of, mtk_dvo_of_ids);
+
+struct platform_driver mtk_dvo_driver = {
+	.probe = mtk_dvo_probe,
+	.remove_new = mtk_dvo_remove,
+	.driver = {
+		.name = "mediatek-dvo",
+		.of_match_table = mtk_dvo_of_ids,
+	},
+};
diff --git a/drivers/gpu/drm/mediatek/mtk_dvo_regs.h b/drivers/gpu/drm/mediatek/mtk_dvo_regs.h
new file mode 100644
index 000000000000..e2c06d0ec225
--- /dev/null
+++ b/drivers/gpu/drm/mediatek/mtk_dvo_regs.h
@@ -0,0 +1,158 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright (c) 2014 MediaTek Inc.
+ * Author: Jie Qiu <jie.qiu@mediatek.com>
+ */
+#ifndef __MTK_DPI_REGS_H
+#define __MTK_DPI_REGS_H
+#define DVO_EN			0x00
+#define EN				BIT(0)
+#define DVO_FORCE_ON			BIT(4)
+#define LINK_OFF			BIT(8)
+#define DVO_RET			0x04
+#define SWRST				BIT(0)
+#define SWRST_SEL			BIT(4)
+#define DVO_INTEN		0x08
+#define INT_VFP_START_EN		BIT(0)
+#define INT_VSYNC_START_EN		BIT(1)
+#define INT_VSYNC_END_EN		BIT(2)
+#define INT_VDE_START_EN		BIT(3)
+#define INT_VDE_END_EN			BIT(4)
+#define INT_WR_INFOQ_REG_EN		BIT(5)
+#define INT_TARGET_LINE0_EN		BIT(6)
+#define INT_TARGET_LINE1_EN		BIT(7)
+#define INT_TARGET_LINE2_EN		BIT(8)
+#define INT_TARGET_LINE3_EN		BIT(9)
+#define INT_WR_INFOQ_START_EN		BIT(10)
+#define INT_WR_INFOQ_END_EN		BIT(11)
+#define EXT_VSYNC_START_EN		BIT(12)
+#define EXT_VSYNC_END_EN		BIT(13)
+#define EXT_VDE_START_EN		BIT(14)
+#define EXT_VDE_END_EN			BIT(15)
+#define EXT_VBLANK_END_EN		BIT(16)
+#define UNDERFLOW_EN			BIT(17)
+#define INFOQ_ABORT_EN			BIT(18)
+#define DVO_INTSTA		0x0c
+#define INT_VFP_START_STA		BIT(0)
+#define INT_VSYNC_START_STA		BIT(1)
+#define INT_VSYNC_END_STA		BIT(2)
+#define INT_VDE_START_STA		BIT(3)
+#define INT_VDE_END_STA			BIT(4)
+#define INT_WR_INFOQ_REG_STA		BIT(5)
+#define INT_TARGET_LINE0_STA		BIT(6)
+#define INT_TARGET_LINE1_STA		BIT(7)
+#define INT_TARGET_LINE2_STA		BIT(8)
+#define INT_TARGET_LINE3_STA		BIT(9)
+#define INT_WR_INFOQ_START_STA		BIT(10)
+#define INT_WR_INFOQ_END_STA		BIT(11)
+#define EXT_VSYNC_START_STA		BIT(12)
+#define EXT_VSYNC_END_STA		BIT(13)
+#define EXT_VDE_START_STA		BIT(14)
+#define EXT_VDE_END_STA			BIT(15)
+#define EXT_VBLANK_END_STA		BIT(16)
+#define INT_UNDERFLOW_STA		BIT(17)
+#define INFOQ_ABORT_STA			BIT(18)
+#define DVO_CON			0x10
+#define INTL_EN				BIT(0)
+#define DVO_OUTPUT_SET		0x18
+#define OUT_NP_SEL			(0x3 << 0)
+#define BIT_SWAP			BIT(4)
+#define CH_SWAP_MASK			(0x7 << 5)
+#define SWAP_RGB			0x00
+#define SWAP_GBR			0x01
+#define SWAP_BRG			0x02
+#define SWAP_RBG			0x03
+#define SWAP_GRB			0x04
+#define SWAP_BGR			0x05
+#define PXL_SWAP			BIT(8)
+#define R_MASK				BIT(12)
+#define G_MASK				BIT(13)
+#define B_MASK				BIT(14)
+#define DE_MASK				BIT(16)
+#define HS_MASK				BIT(17)
+#define VS_MASK				BIT(18)
+#define HS_INV				BIT(19)
+#define VS_INV				BIT(20)
+#define DVO_SRC_SIZE		0x20
+#define SRC_HSIZE			0
+#define SRC_HSIZE_MASK			(0xffff << 0)
+#define SRC_VSIZE			16
+#define SRC_VSIZE_MASK			(0xffff << 16)
+#define DVO_PIC_SIZE		0x24
+#define PIC_HSIZE			0
+#define PIC_HSIZE_MASK			(0xffff << 0)
+#define PIC_VSIZE			16
+#define PIC_VSIZE_MASK			(0xffff << 16)
+#define DVO_TGEN_H0		0x50
+#define HFP				0
+#define HFP_MASK			(0xffff << 0)
+#define HSYNC				16
+#define HSYNC_MASK			(0xffff << 16)
+#define DVO_TGEN_H1		0x54
+#define HSYNC2ACT			0
+#define HSYNC2ACT_MASK			(0xffff << 0)
+#define HACT				16
+#define HACT_MASK			(0xffff << 16)
+#define DVO_TGEN_V0		0x58
+#define VFP				0
+#define VFP_MASK			(0xffff << 0)
+#define VSYNC				16
+#define VSYNC_MASK			(0xffff << 16)
+#define DVO_TGEN_V1		0x5c
+#define VSYNC2ACT			0
+#define VSYNC2ACT_MASK			(0xffff << 0)
+#define VACT				16
+#define VACT_MASK			(0xffff << 16)
+#define DVO_TGEN_INFOQ_LATENCY	0x80
+#define INFOQ_START_LATENCY		0
+#define INFOQ_START_LATENCY_MASK	(0xffff << 0)
+#define INFOQ_END_LATENCY		16
+#define INFOQ_END_LATENCY_MASK		(0xffff << 16)
+#define DVO_BUF_CON0		0x220
+#define DISP_BUF_EN			BIT(0)
+#define FIFO_UNDERFLOW_DONE_BLOCK	BIT(4)
+#define DVO_TGEN_V_LAST_TRAILING_BLANK	0x6c
+#define V_LAST_TRAILING_BLANK			0
+#define V_LAST_TRAILING_BLANK_MASK		(0xffff << 0)
+#define DVO_TGEN_OUTPUT_DELAY_LINE	0x7c
+#define EXT_TG_DLY_LINE				0
+#define EXT_TG_DLY_LINE_MASK			(0xffff << 0)
+#define DVO_MUTEX_VSYNC_SET		0x84
+#define MUTEX_VSYNC_SEL				BIT(0)
+#define MUTEX_VFP_MASK				GENMASK(19, 4)
+#define DVO_PATTERN_CTRL		0x100
+#define PRE_PAT_EN				BIT(0)
+#define PRE_PAT_SEL_MASK			(0x7 << 4)
+#define COLOR_BAR				(0x4<<4)
+#define PRE_PAT_FORCE_ON			BIT(8)
+#define DVO_PATTERN_COLOR		0x104
+#define PAT_R					(0x3ff << 0)
+#define PAT_G					(0x3ff << 10)
+#define PAT_B					(0x3ff << 20)
+#define DVO_SHADOW_CTRL			0x190
+#define FORCE_COMMIT				BIT(0)
+#define BYPASS_SHADOW				BIT(1)
+#define READ_WRK_REG				BIT(2)
+#define DVO_SIZE			0x18
+#define DVO_TGEN_VWIDTH			0x28
+#define DVO_TGEN_VPORCH			0x2c
+#define DVO_TGEN_HPORCH			0x24
+#define DVO_TGEN_HWIDTH			0x20
+#define DVO_BUF_SODI_HIGHT		0x230
+#define DVO_BUF_SODI_LOW		0x234
+#define DVO_BUF_DVFS_HIGHT		0x238
+#define DVO_BUF_DVFS_LOW		0x23c
+#define DVO_BUF_PREULTRA_HIGHT		0x240
+#define DVO_BUF_PREULTRA_LOW		0x244
+#define DVO_BUF_ULTRA_HIGHT		0x248
+#define DVO_BUF_ULTRA_LOW		0x24c
+#define DVO_BUF_URGENT_HIGHT		0x250
+#define DVO_BUF_URGENT_LOW		0x254
+#define DVO_DISP_BUF_MASK		GENMASK(31,0)
+#define DVO_MATRIX_SET			0x140
+#define DVO_CSC_EN				BIT(0)
+#define DVO_INT_MTX_SEL_MASK			GENMASK(8, 4)
+#define DVO_YUV422_SET			0x170
+#define DVO_YUV422_EN				BIT(0)
+#define DVO_CRYCB_MAP				BIT(8)
+#endif /* __MTK_DPI_REGS_H */
-- 
2.45.2

